<html><head><title>GeneValidator - validation explanations</title><meta content="text/html; charset=UTF-8" http-equiv="content-type"><style type="text/css">.lst-kix_euwmeg8cbpt8-4>li:before{content:"\0025cb  "}.lst-kix_g3lepnw8n0x6-4>li:before{content:"\0025cb  "}.lst-kix_9drbe3jtkkuh-4>li:before{content:"" counter(lst-ctn-kix_9drbe3jtkkuh-4,lower-latin) ". "}ul.lst-kix_hzkhz4hhlr8l-6{list-style-type:none}ul.lst-kix_hzkhz4hhlr8l-7{list-style-type:none}ul.lst-kix_hzkhz4hhlr8l-8{list-style-type:none}.lst-kix_z0ttp4mliy4e-6>li:before{content:"\0025cf  "}.lst-kix_tjc8wwk9cseo-2>li:before{content:"\0025a0  "}ol.lst-kix_9drbe3jtkkuh-5.start{counter-reset:lst-ctn-kix_9drbe3jtkkuh-5 0}ul.lst-kix_hzkhz4hhlr8l-0{list-style-type:none}ul.lst-kix_hzkhz4hhlr8l-1{list-style-type:none}ul.lst-kix_hzkhz4hhlr8l-2{list-style-type:none}.lst-kix_tjc8wwk9cseo-3>li:before{content:"\0025cf  "}ul.lst-kix_hzkhz4hhlr8l-3{list-style-type:none}ul.lst-kix_hzkhz4hhlr8l-4{list-style-type:none}ul.lst-kix_hzkhz4hhlr8l-5{list-style-type:none}.lst-kix_dpynbghv8b9j-3>li:before{content:"\0025cf  "}.lst-kix_hzkhz4hhlr8l-6>li:before{content:"\0025cf  "}.lst-kix_7rljtqi4tpf8-2>li:before{content:"\0025a0  "}.lst-kix_yghjnj8s9xuc-2>li:before{content:"\0025a0  "}.lst-kix_cbmcswhr1ftl-6>li:before{content:"\0025cf  "}.lst-kix_dpynbghv8b9j-0>li:before{content:"\0025cf  "}.lst-kix_sjinkw26u0wp-3>li:before{content:"\0025cf  "}.lst-kix_pdbqztwjr65g-8>li:before{content:"\0025a0  "}.lst-kix_qzhxzy2sbaso-7>li:before{content:"\0025cb  "}.lst-kix_3jnvhqhv71nk-4>li:before{content:"\0025cb  "}.lst-kix_fscfg01f3awm-4>li:before{content:"\0025cb  "}.lst-kix_olh7ea290trr-4>li:before{content:"\0025cb  "}.lst-kix_hzkhz4hhlr8l-4>li:before{content:"\0025cb  "}.lst-kix_3jnvhqhv71nk-5>li:before{content:"\0025a0  "}.lst-kix_jamjx369pdr8-2>li:before{content:"\0025a0  "}.lst-kix_x7iifx7mthae-8>li:before{content:"\0025a0  "}.lst-kix_vqtjz6k7exn7-1>li:before{content:"\0025cb  "}.lst-kix_g96zcqu3kh41-8>li:before{content:"\0025a0  "}.lst-kix_gr4a4jnxityy-3>li:before{content:"\0025cf  "}.lst-kix_whzqgipsg1bt-2>li:before{content:"\0025a0  "}.lst-kix_g96zcqu3kh41-0>li:before{content:"\0025cf  "}.lst-kix_puzs83wo6j0o-3>li:before{content:"\0025cf  "}.lst-kix_e5f9ugvogyec-7>li:before{content:"\0025cb  "}.lst-kix_48dltnwvsqfp-1>li:before{content:"\0025cb  "}.lst-kix_4ed4y1s6lj5w-0>li:before{content:"\0025cf  "}.lst-kix_z0ttp4mliy4e-3>li:before{content:"\0025cf  "}.lst-kix_fscfg01f3awm-8>li:before{content:"\0025a0  "}.lst-kix_yghjnj8s9xuc-7>li:before{content:"\0025cb  "}.lst-kix_10avetia2xt4-3>li:before{content:"\0025cf  "}.lst-kix_508olx789l51-3>li:before{content:"\0025cf  "}.lst-kix_10avetia2xt4-8>li:before{content:"\0025a0  "}.lst-kix_9a6zvrglxdu8-1>li:before{content:"\0025cb  "}.lst-kix_z0ttp4mliy4e-7>li:before{content:"\0025cb  "}.lst-kix_vz631vhbp5pc-0>li:before{content:"\0025cf  "}.lst-kix_kmsnp6mmp404-3>li:before{content:"\0025cf  "}.lst-kix_q40ki98g1i63-6>li:before{content:"\0025cf  "}ol.lst-kix_9drbe3jtkkuh-2.start{counter-reset:lst-ctn-kix_9drbe3jtkkuh-2 0}.lst-kix_9drbe3jtkkuh-0>li{counter-increment:lst-ctn-kix_9drbe3jtkkuh-0}.lst-kix_7kppodwmns3b-2>li:before{content:"\0025a0  "}.lst-kix_vqtjz6k7exn7-3>li:before{content:"\0025cf  "}ol.lst-kix_9drbe3jtkkuh-0.start{counter-reset:lst-ctn-kix_9drbe3jtkkuh-0 0}.lst-kix_ao4z72tcnso6-3>li:before{content:"\0025cf  "}.lst-kix_owpom8h3nkwo-7>li:before{content:"\0025cb  "}.lst-kix_7kppodwmns3b-5>li:before{content:"\0025a0  "}.lst-kix_puzs83wo6j0o-2>li:before{content:"\0025a0  "}.lst-kix_e5f9ugvogyec-0>li:before{content:"\0025cf  "}.lst-kix_gr4a4jnxityy-0>li:before{content:"\0025cf  "}ul.lst-kix_dpynbghv8b9j-8{list-style-type:none}.lst-kix_euwmeg8cbpt8-6>li:before{content:"\0025cf  "}ul.lst-kix_dpynbghv8b9j-6{list-style-type:none}ul.lst-kix_dpynbghv8b9j-7{list-style-type:none}ul.lst-kix_dpynbghv8b9j-4{list-style-type:none}ul.lst-kix_dpynbghv8b9j-5{list-style-type:none}ul.lst-kix_dpynbghv8b9j-2{list-style-type:none}ul.lst-kix_dpynbghv8b9j-3{list-style-type:none}ul.lst-kix_puzs83wo6j0o-6{list-style-type:none}ul.lst-kix_dpynbghv8b9j-1{list-style-type:none}ul.lst-kix_puzs83wo6j0o-5{list-style-type:none}ul.lst-kix_dpynbghv8b9j-0{list-style-type:none}ul.lst-kix_puzs83wo6j0o-8{list-style-type:none}ul.lst-kix_puzs83wo6j0o-7{list-style-type:none}ul.lst-kix_puzs83wo6j0o-0{list-style-type:none}.lst-kix_pdbqztwjr65g-5>li:before{content:"\0025a0  "}ul.lst-kix_puzs83wo6j0o-2{list-style-type:none}.lst-kix_myo05blafbk-5>li:before{content:"\0025a0  "}ul.lst-kix_puzs83wo6j0o-1{list-style-type:none}ul.lst-kix_puzs83wo6j0o-4{list-style-type:none}ul.lst-kix_puzs83wo6j0o-3{list-style-type:none}.lst-kix_48dltnwvsqfp-5>li:before{content:"\0025a0  "}.lst-kix_e5f9ugvogyec-8>li:before{content:"\0025a0  "}.lst-kix_fscfg01f3awm-0>li:before{content:"\0025cf  "}.lst-kix_riib1nssx348-5>li:before{content:"\0025a0  "}.lst-kix_508olx789l51-4>li:before{content:"\0025cb  "}.lst-kix_48dltnwvsqfp-7>li:before{content:"\0025cb  "}.lst-kix_10avetia2xt4-4>li:before{content:"\0025cb  "}.lst-kix_7kppodwmns3b-6>li:before{content:"\0025cf  "}.lst-kix_48dltnwvsqfp-3>li:before{content:"\0025cf  "}.lst-kix_4tq1geabcxa-2>li:before{content:"\0025a0  "}ol.lst-kix_9drbe3jtkkuh-0{list-style-type:none}.lst-kix_qzhxzy2sbaso-1>li:before{content:"\0025cb  "}.lst-kix_7rljtqi4tpf8-3>li:before{content:"\0025cf  "}.lst-kix_4tq1geabcxa-7>li:before{content:"\0025cb  "}.lst-kix_puzs83wo6j0o-5>li:before{content:"\0025a0  "}.lst-kix_b24v03uyl4k9-4>li:before{content:"\0025cb  "}.lst-kix_velyrizglpvf-6>li:before{content:"\0025cf  "}.lst-kix_sjinkw26u0wp-5>li:before{content:"\0025a0  "}.lst-kix_xc244bhq3ube-8>li:before{content:"\0025a0  "}.lst-kix_hzkhz4hhlr8l-1>li:before{content:"\0025cb  "}.lst-kix_l7t8lma3v5dd-7>li:before{content:"\0025cb  "}.lst-kix_jamjx369pdr8-5>li:before{content:"\0025a0  "}.lst-kix_vqtjz6k7exn7-0>li:before{content:"\0025cf  "}.lst-kix_e5f9ugvogyec-1>li:before{content:"\0025cb  "}.lst-kix_4ed4y1s6lj5w-7>li:before{content:"\0025cb  "}.lst-kix_olh7ea290trr-7>li:before{content:"\0025cb  "}.lst-kix_48dltnwvsqfp-6>li:before{content:"\0025cf  "}.lst-kix_velyrizglpvf-0>li:before{content:"\0025cf  "}.lst-kix_mk8t5sy0k589-2>li:before{content:"\0025a0  "}.lst-kix_yghjnj8s9xuc-6>li:before{content:"\0025cf  "}.lst-kix_b24v03uyl4k9-0>li:before{content:"\0025cf  "}.lst-kix_mk8t5sy0k589-4>li:before{content:"\0025cb  "}.lst-kix_g96zcqu3kh41-6>li:before{content:"\0025cf  "}.lst-kix_whzqgipsg1bt-3>li:before{content:"\0025cf  "}.lst-kix_euwmeg8cbpt8-0>li:before{content:"\0025cf  "}.lst-kix_olh7ea290trr-6>li:before{content:"\0025cf  "}ol.lst-kix_9drbe3jtkkuh-3{list-style-type:none}ol.lst-kix_9drbe3jtkkuh-4{list-style-type:none}.lst-kix_g3lepnw8n0x6-5>li:before{content:"\0025a0  "}ol.lst-kix_9drbe3jtkkuh-1{list-style-type:none}ol.lst-kix_9drbe3jtkkuh-2{list-style-type:none}ol.lst-kix_9drbe3jtkkuh-7{list-style-type:none}ol.lst-kix_9drbe3jtkkuh-8{list-style-type:none}ol.lst-kix_9drbe3jtkkuh-5{list-style-type:none}.lst-kix_owpom8h3nkwo-3>li:before{content:"\0025cf  "}ol.lst-kix_9drbe3jtkkuh-6{list-style-type:none}.lst-kix_gr4a4jnxityy-8>li:before{content:"\0025a0  "}ul.lst-kix_g96zcqu3kh41-1{list-style-type:none}.lst-kix_dpynbghv8b9j-5>li:before{content:"\0025a0  "}ul.lst-kix_g96zcqu3kh41-2{list-style-type:none}ul.lst-kix_g96zcqu3kh41-3{list-style-type:none}ul.lst-kix_g96zcqu3kh41-4{list-style-type:none}ul.lst-kix_jamjx369pdr8-8{list-style-type:none}ul.lst-kix_jamjx369pdr8-7{list-style-type:none}.lst-kix_7rljtqi4tpf8-8>li:before{content:"\0025a0  "}ul.lst-kix_jamjx369pdr8-6{list-style-type:none}ul.lst-kix_jamjx369pdr8-5{list-style-type:none}ul.lst-kix_g96zcqu3kh41-0{list-style-type:none}.lst-kix_ao4z72tcnso6-1>li:before{content:"\0025cb  "}.lst-kix_g3lepnw8n0x6-0>li:before{content:"\0025cf  "}.lst-kix_x7iifx7mthae-0>li:before{content:"\0025cf  "}.lst-kix_mk8t5sy0k589-5>li:before{content:"\0025a0  "}.lst-kix_q40ki98g1i63-7>li:before{content:"\0025cb  "}.lst-kix_10avetia2xt4-6>li:before{content:"\0025cf  "}.lst-kix_9drbe3jtkkuh-6>li:before{content:"" counter(lst-ctn-kix_9drbe3jtkkuh-6,decimal) ". "}.lst-kix_cmqbbecr9wqr-2>li:before{content:"\0025a0  "}ul.lst-kix_jamjx369pdr8-3{list-style-type:none}.lst-kix_l7t8lma3v5dd-3>li:before{content:"\0025cf  "}ul.lst-kix_jamjx369pdr8-4{list-style-type:none}ul.lst-kix_jamjx369pdr8-1{list-style-type:none}ul.lst-kix_jamjx369pdr8-2{list-style-type:none}ul.lst-kix_jamjx369pdr8-0{list-style-type:none}.lst-kix_9drbe3jtkkuh-3>li:before{content:"" counter(lst-ctn-kix_9drbe3jtkkuh-3,decimal) ". "}.lst-kix_7kppodwmns3b-3>li:before{content:"\0025cf  "}.lst-kix_sjinkw26u0wp-1>li:before{content:"\0025cb  "}.lst-kix_tjc8wwk9cseo-4>li:before{content:"\0025cb  "}ul.lst-kix_sjinkw26u0wp-0{list-style-type:none}.lst-kix_7rljtqi4tpf8-6>li:before{content:"\0025cf  "}.lst-kix_l7t8lma3v5dd-5>li:before{content:"\0025a0  "}.lst-kix_4ed4y1s6lj5w-8>li:before{content:"\0025a0  "}ul.lst-kix_g3lepnw8n0x6-0{list-style-type:none}ul.lst-kix_g3lepnw8n0x6-1{list-style-type:none}ul.lst-kix_g3lepnw8n0x6-2{list-style-type:none}ul.lst-kix_g3lepnw8n0x6-3{list-style-type:none}ul.lst-kix_g3lepnw8n0x6-4{list-style-type:none}ul.lst-kix_g3lepnw8n0x6-5{list-style-type:none}ul.lst-kix_g3lepnw8n0x6-6{list-style-type:none}ul.lst-kix_g3lepnw8n0x6-7{list-style-type:none}.lst-kix_mk8t5sy0k589-3>li:before{content:"\0025cf  "}ul.lst-kix_g3lepnw8n0x6-8{list-style-type:none}.lst-kix_jamjx369pdr8-4>li:before{content:"\0025cb  "}.lst-kix_xc244bhq3ube-7>li:before{content:"\0025cb  "}.lst-kix_48dltnwvsqfp-8>li:before{content:"\0025a0  "}.lst-kix_puzs83wo6j0o-4>li:before{content:"\0025cb  "}.lst-kix_9drbe3jtkkuh-5>li:before{content:"" counter(lst-ctn-kix_9drbe3jtkkuh-5,lower-roman) ". "}.lst-kix_vz631vhbp5pc-8>li:before{content:"\0025a0  "}.lst-kix_mk8t5sy0k589-6>li:before{content:"\0025cf  "}.lst-kix_508olx789l51-1>li:before{content:"\0025cb  "}ul.lst-kix_sjinkw26u0wp-8{list-style-type:none}.lst-kix_sjinkw26u0wp-6>li:before{content:"\0025cf  "}ul.lst-kix_sjinkw26u0wp-7{list-style-type:none}ul.lst-kix_sjinkw26u0wp-6{list-style-type:none}.lst-kix_owpom8h3nkwo-4>li:before{content:"\0025cb  "}ul.lst-kix_sjinkw26u0wp-5{list-style-type:none}.lst-kix_ao4z72tcnso6-6>li:before{content:"\0025cf  "}ul.lst-kix_sjinkw26u0wp-4{list-style-type:none}ul.lst-kix_sjinkw26u0wp-3{list-style-type:none}ul.lst-kix_sjinkw26u0wp-2{list-style-type:none}.lst-kix_gr4a4jnxityy-1>li:before{content:"\0025cb  "}ul.lst-kix_sjinkw26u0wp-1{list-style-type:none}.lst-kix_hzkhz4hhlr8l-3>li:before{content:"\0025cf  "}.lst-kix_cmqbbecr9wqr-6>li:before{content:"\0025cf  "}.lst-kix_whzqgipsg1bt-6>li:before{content:"\0025cf  "}.lst-kix_x7t5b7jbczs4-7>li:before{content:"\0025cb  "}.lst-kix_pdbqztwjr65g-7>li:before{content:"\0025cb  "}.lst-kix_x7t5b7jbczs4-2>li:before{content:"\0025a0  "}.lst-kix_7kppodwmns3b-1>li:before{content:"\0025cb  "}.lst-kix_velyrizglpvf-1>li:before{content:"\0025cb  "}.lst-kix_x7iifx7mthae-6>li:before{content:"\0025cf  "}ul.lst-kix_l7t8lma3v5dd-7{list-style-type:none}ul.lst-kix_l7t8lma3v5dd-6{list-style-type:none}ul.lst-kix_l7t8lma3v5dd-5{list-style-type:none}.lst-kix_dpynbghv8b9j-7>li:before{content:"\0025cb  "}ul.lst-kix_l7t8lma3v5dd-4{list-style-type:none}ul.lst-kix_l7t8lma3v5dd-3{list-style-type:none}ul.lst-kix_l7t8lma3v5dd-2{list-style-type:none}ul.lst-kix_l7t8lma3v5dd-1{list-style-type:none}ul.lst-kix_l7t8lma3v5dd-0{list-style-type:none}.lst-kix_whzqgipsg1bt-5>li:before{content:"\0025a0  "}.lst-kix_l7t8lma3v5dd-2>li:before{content:"\0025a0  "}.lst-kix_riib1nssx348-8>li:before{content:"\0025a0  "}ul.lst-kix_l7t8lma3v5dd-8{list-style-type:none}.lst-kix_myo05blafbk-7>li:before{content:"\0025cb  "}.lst-kix_vz631vhbp5pc-2>li:before{content:"\0025a0  "}.lst-kix_cbmcswhr1ftl-0>li:before{content:"\0025cf  "}.lst-kix_qzhxzy2sbaso-4>li:before{content:"\0025cb  "}.lst-kix_3jnvhqhv71nk-7>li:before{content:"\0025cb  "}.lst-kix_ao4z72tcnso6-0>li:before{content:"\0025cf  "}.lst-kix_3jnvhqhv71nk-6>li:before{content:"\0025cf  "}.lst-kix_x7t5b7jbczs4-5>li:before{content:"\0025a0  "}.lst-kix_mk8t5sy0k589-8>li:before{content:"\0025a0  "}.lst-kix_owpom8h3nkwo-8>li:before{content:"\0025a0  "}.lst-kix_mk8t5sy0k589-1>li:before{content:"\0025cb  "}.lst-kix_3kgcjgrfg4nh-2>li:before{content:"\0025a0  "}.lst-kix_7kppodwmns3b-4>li:before{content:"\0025cb  "}.lst-kix_b24v03uyl4k9-8>li:before{content:"\0025a0  "}.lst-kix_whzqgipsg1bt-8>li:before{content:"\0025a0  "}ul.lst-kix_4tq1geabcxa-8{list-style-type:none}ul.lst-kix_4tq1geabcxa-6{list-style-type:none}ul.lst-kix_4tq1geabcxa-7{list-style-type:none}ul.lst-kix_4tq1geabcxa-4{list-style-type:none}ul.lst-kix_4tq1geabcxa-5{list-style-type:none}ul.lst-kix_4tq1geabcxa-2{list-style-type:none}ul.lst-kix_4tq1geabcxa-3{list-style-type:none}.lst-kix_myo05blafbk-3>li:before{content:"\0025cf  "}ul.lst-kix_4tq1geabcxa-0{list-style-type:none}ul.lst-kix_4tq1geabcxa-1{list-style-type:none}ul.lst-kix_mk8t5sy0k589-3{list-style-type:none}ul.lst-kix_mk8t5sy0k589-4{list-style-type:none}ul.lst-kix_mk8t5sy0k589-5{list-style-type:none}.lst-kix_myo05blafbk-0>li:before{content:"\0025cf  "}ul.lst-kix_z0ttp4mliy4e-8{list-style-type:none}ul.lst-kix_mk8t5sy0k589-6{list-style-type:none}.lst-kix_myo05blafbk-6>li:before{content:"\0025cf  "}ul.lst-kix_z0ttp4mliy4e-7{list-style-type:none}ul.lst-kix_mk8t5sy0k589-0{list-style-type:none}.lst-kix_48dltnwvsqfp-2>li:before{content:"\0025a0  "}ul.lst-kix_mk8t5sy0k589-1{list-style-type:none}ul.lst-kix_mk8t5sy0k589-2{list-style-type:none}.lst-kix_3kgcjgrfg4nh-7>li:before{content:"\0025cb  "}ul.lst-kix_yghjnj8s9xuc-7{list-style-type:none}ul.lst-kix_fscfg01f3awm-2{list-style-type:none}ul.lst-kix_yghjnj8s9xuc-8{list-style-type:none}ul.lst-kix_fscfg01f3awm-1{list-style-type:none}ul.lst-kix_yghjnj8s9xuc-5{list-style-type:none}ul.lst-kix_fscfg01f3awm-0{list-style-type:none}ul.lst-kix_yghjnj8s9xuc-6{list-style-type:none}.lst-kix_q40ki98g1i63-3>li:before{content:"\0025cf  "}ul.lst-kix_fscfg01f3awm-6{list-style-type:none}ul.lst-kix_fscfg01f3awm-5{list-style-type:none}ul.lst-kix_fscfg01f3awm-4{list-style-type:none}.lst-kix_vqtjz6k7exn7-5>li:before{content:"\0025a0  "}.lst-kix_9a6zvrglxdu8-2>li:before{content:"\0025a0  "}ul.lst-kix_fscfg01f3awm-3{list-style-type:none}.lst-kix_b24v03uyl4k9-1>li:before{content:"\0025cb  "}.lst-kix_508olx789l51-0>li:before{content:"\0025cf  "}ul.lst-kix_z0ttp4mliy4e-0{list-style-type:none}ul.lst-kix_yghjnj8s9xuc-0{list-style-type:none}ul.lst-kix_z0ttp4mliy4e-1{list-style-type:none}ul.lst-kix_fscfg01f3awm-8{list-style-type:none}ul.lst-kix_z0ttp4mliy4e-2{list-style-type:none}ul.lst-kix_fscfg01f3awm-7{list-style-type:none}ul.lst-kix_z0ttp4mliy4e-3{list-style-type:none}ul.lst-kix_yghjnj8s9xuc-3{list-style-type:none}ul.lst-kix_z0ttp4mliy4e-4{list-style-type:none}ul.lst-kix_yghjnj8s9xuc-4{list-style-type:none}ul.lst-kix_mk8t5sy0k589-8{list-style-type:none}ul.lst-kix_z0ttp4mliy4e-5{list-style-type:none}ul.lst-kix_yghjnj8s9xuc-1{list-style-type:none}ul.lst-kix_mk8t5sy0k589-7{list-style-type:none}ul.lst-kix_z0ttp4mliy4e-6{list-style-type:none}ul.lst-kix_yghjnj8s9xuc-2{list-style-type:none}.lst-kix_9a6zvrglxdu8-3>li:before{content:"\0025cf  "}.lst-kix_x7iifx7mthae-5>li:before{content:"\0025a0  "}.lst-kix_3jnvhqhv71nk-3>li:before{content:"\0025cf  "}ul.lst-kix_7rljtqi4tpf8-0{list-style-type:none}.lst-kix_yghjnj8s9xuc-1>li:before{content:"\0025cb  "}ul.lst-kix_7rljtqi4tpf8-1{list-style-type:none}ul.lst-kix_7rljtqi4tpf8-3{list-style-type:none}.lst-kix_sjinkw26u0wp-4>li:before{content:"\0025cb  "}ul.lst-kix_q40ki98g1i63-6{list-style-type:none}ul.lst-kix_7rljtqi4tpf8-2{list-style-type:none}ul.lst-kix_q40ki98g1i63-5{list-style-type:none}ul.lst-kix_7rljtqi4tpf8-5{list-style-type:none}ul.lst-kix_q40ki98g1i63-4{list-style-type:none}ul.lst-kix_7rljtqi4tpf8-4{list-style-type:none}ul.lst-kix_q40ki98g1i63-3{list-style-type:none}ul.lst-kix_7rljtqi4tpf8-7{list-style-type:none}ul.lst-kix_q40ki98g1i63-2{list-style-type:none}ul.lst-kix_7rljtqi4tpf8-6{list-style-type:none}ul.lst-kix_q40ki98g1i63-1{list-style-type:none}ul.lst-kix_q40ki98g1i63-0{list-style-type:none}ul.lst-kix_7rljtqi4tpf8-8{list-style-type:none}.lst-kix_qzhxzy2sbaso-0>li:before{content:"\0025cf  "}ul.lst-kix_q40ki98g1i63-8{list-style-type:none}.lst-kix_g96zcqu3kh41-7>li:before{content:"\0025cb  "}ul.lst-kix_q40ki98g1i63-7{list-style-type:none}ul.lst-kix_e5f9ugvogyec-8{list-style-type:none}ul.lst-kix_cmqbbecr9wqr-3{list-style-type:none}ul.lst-kix_e5f9ugvogyec-7{list-style-type:none}ul.lst-kix_cmqbbecr9wqr-4{list-style-type:none}.lst-kix_mk8t5sy0k589-7>li:before{content:"\0025cb  "}ul.lst-kix_cmqbbecr9wqr-1{list-style-type:none}ul.lst-kix_cmqbbecr9wqr-2{list-style-type:none}ul.lst-kix_e5f9ugvogyec-4{list-style-type:none}.lst-kix_10avetia2xt4-2>li:before{content:"\0025a0  "}ul.lst-kix_e5f9ugvogyec-3{list-style-type:none}ul.lst-kix_cmqbbecr9wqr-0{list-style-type:none}ul.lst-kix_e5f9ugvogyec-6{list-style-type:none}ul.lst-kix_e5f9ugvogyec-5{list-style-type:none}.lst-kix_vz631vhbp5pc-6>li:before{content:"\0025cf  "}.lst-kix_xc244bhq3ube-4>li:before{content:"\0025cb  "}.lst-kix_kmsnp6mmp404-4>li:before{content:"\0025cb  "}.lst-kix_b24v03uyl4k9-7>li:before{content:"\0025cb  "}.lst-kix_b24v03uyl4k9-6>li:before{content:"\0025cf  "}ol.lst-kix_9drbe3jtkkuh-4.start{counter-reset:lst-ctn-kix_9drbe3jtkkuh-4 0}.lst-kix_b24v03uyl4k9-5>li:before{content:"\0025a0  "}.lst-kix_4ed4y1s6lj5w-6>li:before{content:"\0025cf  "}.lst-kix_e5f9ugvogyec-6>li:before{content:"\0025cf  "}ul.lst-kix_cmqbbecr9wqr-8{list-style-type:none}ul.lst-kix_cmqbbecr9wqr-7{list-style-type:none}.lst-kix_508olx789l51-6>li:before{content:"\0025cf  "}ul.lst-kix_cmqbbecr9wqr-6{list-style-type:none}ul.lst-kix_cmqbbecr9wqr-5{list-style-type:none}.lst-kix_4tq1geabcxa-5>li:before{content:"\0025a0  "}.lst-kix_3jnvhqhv71nk-2>li:before{content:"\0025a0  "}ol.lst-kix_9drbe3jtkkuh-6.start{counter-reset:lst-ctn-kix_9drbe3jtkkuh-6 0}.lst-kix_velyrizglpvf-7>li:before{content:"\0025cb  "}ul.lst-kix_qzhxzy2sbaso-7{list-style-type:none}ul.lst-kix_qzhxzy2sbaso-8{list-style-type:none}.lst-kix_9drbe3jtkkuh-8>li:before{content:"" counter(lst-ctn-kix_9drbe3jtkkuh-8,lower-roman) ". "}ul.lst-kix_qzhxzy2sbaso-5{list-style-type:none}ul.lst-kix_qzhxzy2sbaso-6{list-style-type:none}.lst-kix_b24v03uyl4k9-2>li:before{content:"\0025a0  "}ul.lst-kix_qzhxzy2sbaso-0{list-style-type:none}.lst-kix_4tq1geabcxa-1>li:before{content:"\0025cb  "}ul.lst-kix_e5f9ugvogyec-1{list-style-type:none}ul.lst-kix_qzhxzy2sbaso-3{list-style-type:none}ul.lst-kix_e5f9ugvogyec-2{list-style-type:none}ul.lst-kix_qzhxzy2sbaso-4{list-style-type:none}ul.lst-kix_qzhxzy2sbaso-1{list-style-type:none}ul.lst-kix_e5f9ugvogyec-0{list-style-type:none}ul.lst-kix_qzhxzy2sbaso-2{list-style-type:none}.lst-kix_cbmcswhr1ftl-5>li:before{content:"\0025a0  "}.lst-kix_q40ki98g1i63-5>li:before{content:"\0025a0  "}.lst-kix_kmsnp6mmp404-1>li:before{content:"\0025cb  "}.lst-kix_z0ttp4mliy4e-2>li:before{content:"\0025a0  "}.lst-kix_olh7ea290trr-0>li:before{content:"\0025cf  "}.lst-kix_x7t5b7jbczs4-4>li:before{content:"\0025cb  "}.lst-kix_ao4z72tcnso6-8>li:before{content:"\0025a0  "}.lst-kix_9drbe3jtkkuh-1>li:before{content:"" counter(lst-ctn-kix_9drbe3jtkkuh-1,lower-latin) ". "}.lst-kix_gr4a4jnxityy-7>li:before{content:"\0025cb  "}.lst-kix_x7t5b7jbczs4-1>li:before{content:"\0025cb  "}.lst-kix_whzqgipsg1bt-7>li:before{content:"\0025cb  "}.lst-kix_9drbe3jtkkuh-1>li{counter-increment:lst-ctn-kix_9drbe3jtkkuh-1}.lst-kix_mk8t5sy0k589-0>li:before{content:"\0025cf  "}.lst-kix_jamjx369pdr8-7>li:before{content:"\0025cb  "}.lst-kix_ao4z72tcnso6-2>li:before{content:"\0025a0  "}.lst-kix_48dltnwvsqfp-4>li:before{content:"\0025cb  "}.lst-kix_g96zcqu3kh41-3>li:before{content:"\0025cf  "}.lst-kix_9drbe3jtkkuh-4>li{counter-increment:lst-ctn-kix_9drbe3jtkkuh-4}.lst-kix_fscfg01f3awm-6>li:before{content:"\0025cf  "}.lst-kix_4ed4y1s6lj5w-2>li:before{content:"\0025a0  "}.lst-kix_kmsnp6mmp404-5>li:before{content:"\0025a0  "}.lst-kix_vz631vhbp5pc-5>li:before{content:"\0025a0  "}.lst-kix_qzhxzy2sbaso-2>li:before{content:"\0025a0  "}.lst-kix_x7t5b7jbczs4-6>li:before{content:"\0025cf  "}.lst-kix_9drbe3jtkkuh-6>li{counter-increment:lst-ctn-kix_9drbe3jtkkuh-6}ul.lst-kix_g96zcqu3kh41-5{list-style-type:none}ul.lst-kix_g96zcqu3kh41-6{list-style-type:none}ul.lst-kix_g96zcqu3kh41-7{list-style-type:none}.lst-kix_euwmeg8cbpt8-2>li:before{content:"\0025a0  "}ul.lst-kix_g96zcqu3kh41-8{list-style-type:none}.lst-kix_10avetia2xt4-1>li:before{content:"\0025cb  "}.lst-kix_myo05blafbk-4>li:before{content:"\0025cb  "}.lst-kix_tjc8wwk9cseo-8>li:before{content:"\0025a0  "}.lst-kix_9a6zvrglxdu8-8>li:before{content:"\0025a0  "}.lst-kix_9a6zvrglxdu8-4>li:before{content:"\0025cb  "}.lst-kix_yghjnj8s9xuc-0>li:before{content:"\0025cf  "}.lst-kix_x7iifx7mthae-7>li:before{content:"\0025cb  "}.lst-kix_vz631vhbp5pc-3>li:before{content:"\0025cf  "}.lst-kix_g96zcqu3kh41-1>li:before{content:"\0025cb  "}.lst-kix_7rljtqi4tpf8-5>li:before{content:"\0025a0  "}.lst-kix_tjc8wwk9cseo-5>li:before{content:"\0025a0  "}.lst-kix_sjinkw26u0wp-2>li:before{content:"\0025a0  "}.lst-kix_4tq1geabcxa-4>li:before{content:"\0025cb  "}.lst-kix_q40ki98g1i63-8>li:before{content:"\0025a0  "}ul.lst-kix_gr4a4jnxityy-2{list-style-type:none}ul.lst-kix_gr4a4jnxityy-3{list-style-type:none}ul.lst-kix_gr4a4jnxityy-4{list-style-type:none}ul.lst-kix_gr4a4jnxityy-5{list-style-type:none}ul.lst-kix_gr4a4jnxityy-0{list-style-type:none}ul.lst-kix_gr4a4jnxityy-1{list-style-type:none}.lst-kix_owpom8h3nkwo-1>li:before{content:"\0025cb  "}ul.lst-kix_gr4a4jnxityy-6{list-style-type:none}ul.lst-kix_gr4a4jnxityy-7{list-style-type:none}.lst-kix_9drbe3jtkkuh-7>li{counter-increment:lst-ctn-kix_9drbe3jtkkuh-7}ul.lst-kix_gr4a4jnxityy-8{list-style-type:none}.lst-kix_48dltnwvsqfp-0>li:before{content:"\0025cf  "}.lst-kix_kmsnp6mmp404-7>li:before{content:"\0025cb  "}ul.lst-kix_vz631vhbp5pc-2{list-style-type:none}ul.lst-kix_vz631vhbp5pc-1{list-style-type:none}ul.lst-kix_vz631vhbp5pc-4{list-style-type:none}ul.lst-kix_vz631vhbp5pc-3{list-style-type:none}ul.lst-kix_vz631vhbp5pc-6{list-style-type:none}ul.lst-kix_vz631vhbp5pc-5{list-style-type:none}ul.lst-kix_vz631vhbp5pc-8{list-style-type:none}ul.lst-kix_vz631vhbp5pc-7{list-style-type:none}ul.lst-kix_vz631vhbp5pc-0{list-style-type:none}.lst-kix_jamjx369pdr8-3>li:before{content:"\0025cf  "}.lst-kix_xc244bhq3ube-1>li:before{content:"\0025cb  "}ul.lst-kix_ao4z72tcnso6-5{list-style-type:none}ul.lst-kix_ao4z72tcnso6-6{list-style-type:none}ul.lst-kix_ao4z72tcnso6-3{list-style-type:none}ul.lst-kix_ao4z72tcnso6-4{list-style-type:none}.lst-kix_owpom8h3nkwo-6>li:before{content:"\0025cf  "}ul.lst-kix_ao4z72tcnso6-1{list-style-type:none}ul.lst-kix_ao4z72tcnso6-2{list-style-type:none}ul.lst-kix_ao4z72tcnso6-0{list-style-type:none}.lst-kix_yghjnj8s9xuc-4>li:before{content:"\0025cb  "}.lst-kix_fscfg01f3awm-3>li:before{content:"\0025cf  "}ul.lst-kix_ao4z72tcnso6-8{list-style-type:none}ul.lst-kix_ao4z72tcnso6-7{list-style-type:none}.lst-kix_fscfg01f3awm-1>li:before{content:"\0025cb  "}.lst-kix_cbmcswhr1ftl-1>li:before{content:"\0025cb  "}.lst-kix_vz631vhbp5pc-1>li:before{content:"\0025cb  "}.lst-kix_owpom8h3nkwo-2>li:before{content:"\0025a0  "}.lst-kix_vqtjz6k7exn7-7>li:before{content:"\0025cb  "}ul.lst-kix_myo05blafbk-1{list-style-type:none}ul.lst-kix_myo05blafbk-2{list-style-type:none}ul.lst-kix_myo05blafbk-3{list-style-type:none}ul.lst-kix_myo05blafbk-4{list-style-type:none}ul.lst-kix_myo05blafbk-5{list-style-type:none}ul.lst-kix_myo05blafbk-6{list-style-type:none}.lst-kix_dpynbghv8b9j-6>li:before{content:"\0025cf  "}ul.lst-kix_myo05blafbk-7{list-style-type:none}ul.lst-kix_myo05blafbk-8{list-style-type:none}.lst-kix_gr4a4jnxityy-2>li:before{content:"\0025a0  "}ul.lst-kix_myo05blafbk-0{list-style-type:none}.lst-kix_tjc8wwk9cseo-0>li:before{content:"\0025cf  "}.lst-kix_puzs83wo6j0o-7>li:before{content:"\0025cb  "}.lst-kix_fscfg01f3awm-2>li:before{content:"\0025a0  "}.lst-kix_z0ttp4mliy4e-8>li:before{content:"\0025a0  "}.lst-kix_euwmeg8cbpt8-3>li:before{content:"\0025cf  "}.lst-kix_sjinkw26u0wp-7>li:before{content:"\0025cb  "}.lst-kix_gr4a4jnxityy-6>li:before{content:"\0025cf  "}.lst-kix_l7t8lma3v5dd-1>li:before{content:"\0025cb  "}.lst-kix_xc244bhq3ube-3>li:before{content:"\0025cf  "}.lst-kix_xc244bhq3ube-2>li:before{content:"\0025a0  "}.lst-kix_x7t5b7jbczs4-8>li:before{content:"\0025a0  "}.lst-kix_qzhxzy2sbaso-6>li:before{content:"\0025cf  "}.lst-kix_hzkhz4hhlr8l-0>li:before{content:"\0025cf  "}ul.lst-kix_4ed4y1s6lj5w-8{list-style-type:none}ul.lst-kix_4ed4y1s6lj5w-7{list-style-type:none}.lst-kix_z0ttp4mliy4e-0>li:before{content:"\0025cf  "}ul.lst-kix_4ed4y1s6lj5w-6{list-style-type:none}ul.lst-kix_4ed4y1s6lj5w-5{list-style-type:none}.lst-kix_q40ki98g1i63-4>li:before{content:"\0025cb  "}.lst-kix_e5f9ugvogyec-3>li:before{content:"\0025cf  "}ul.lst-kix_whzqgipsg1bt-3{list-style-type:none}ul.lst-kix_whzqgipsg1bt-2{list-style-type:none}ul.lst-kix_whzqgipsg1bt-1{list-style-type:none}ul.lst-kix_whzqgipsg1bt-0{list-style-type:none}ul.lst-kix_whzqgipsg1bt-7{list-style-type:none}ul.lst-kix_whzqgipsg1bt-6{list-style-type:none}ul.lst-kix_whzqgipsg1bt-5{list-style-type:none}ul.lst-kix_whzqgipsg1bt-4{list-style-type:none}.lst-kix_puzs83wo6j0o-0>li:before{content:"\0025cf  "}.lst-kix_x7iifx7mthae-1>li:before{content:"\0025cb  "}ul.lst-kix_whzqgipsg1bt-8{list-style-type:none}.lst-kix_9a6zvrglxdu8-6>li:before{content:"\0025cf  "}.lst-kix_x7t5b7jbczs4-0>li:before{content:"\0025cf  "}.lst-kix_pdbqztwjr65g-4>li:before{content:"\0025cb  "}.lst-kix_7kppodwmns3b-0>li:before{content:"\0025cf  "}.lst-kix_508olx789l51-5>li:before{content:"\0025a0  "}.lst-kix_vz631vhbp5pc-4>li:before{content:"\0025cb  "}.lst-kix_9a6zvrglxdu8-0>li:before{content:"\0025cf  "}.lst-kix_10avetia2xt4-7>li:before{content:"\0025cb  "}.lst-kix_q40ki98g1i63-0>li:before{content:"\0025cf  "}.lst-kix_3kgcjgrfg4nh-3>li:before{content:"\0025cf  "}ul.lst-kix_riib1nssx348-3{list-style-type:none}ul.lst-kix_riib1nssx348-4{list-style-type:none}ul.lst-kix_riib1nssx348-1{list-style-type:none}ul.lst-kix_riib1nssx348-2{list-style-type:none}ul.lst-kix_riib1nssx348-0{list-style-type:none}.lst-kix_tjc8wwk9cseo-1>li:before{content:"\0025cb  "}.lst-kix_ao4z72tcnso6-4>li:before{content:"\0025cb  "}.lst-kix_velyrizglpvf-5>li:before{content:"\0025a0  "}ul.lst-kix_olh7ea290trr-8{list-style-type:none}ul.lst-kix_olh7ea290trr-7{list-style-type:none}ul.lst-kix_olh7ea290trr-6{list-style-type:none}ul.lst-kix_olh7ea290trr-5{list-style-type:none}ul.lst-kix_olh7ea290trr-4{list-style-type:none}.lst-kix_9drbe3jtkkuh-8>li{counter-increment:lst-ctn-kix_9drbe3jtkkuh-8}ul.lst-kix_olh7ea290trr-3{list-style-type:none}ul.lst-kix_olh7ea290trr-2{list-style-type:none}ul.lst-kix_olh7ea290trr-1{list-style-type:none}ul.lst-kix_riib1nssx348-8{list-style-type:none}ul.lst-kix_olh7ea290trr-0{list-style-type:none}ul.lst-kix_riib1nssx348-7{list-style-type:none}ul.lst-kix_riib1nssx348-6{list-style-type:none}ul.lst-kix_riib1nssx348-5{list-style-type:none}.lst-kix_puzs83wo6j0o-8>li:before{content:"\0025a0  "}.lst-kix_q40ki98g1i63-1>li:before{content:"\0025cb  "}.lst-kix_4ed4y1s6lj5w-1>li:before{content:"\0025cb  "}.lst-kix_pdbqztwjr65g-6>li:before{content:"\0025cf  "}.lst-kix_cbmcswhr1ftl-2>li:before{content:"\0025a0  "}.lst-kix_euwmeg8cbpt8-8>li:before{content:"\0025a0  "}.lst-kix_7kppodwmns3b-8>li:before{content:"\0025a0  "}.lst-kix_euwmeg8cbpt8-7>li:before{content:"\0025cb  "}ol.lst-kix_9drbe3jtkkuh-3.start{counter-reset:lst-ctn-kix_9drbe3jtkkuh-3 0}ul.lst-kix_x7t5b7jbczs4-8{list-style-type:none}.lst-kix_3kgcjgrfg4nh-1>li:before{content:"\0025cb  "}.lst-kix_cmqbbecr9wqr-0>li:before{content:"\0025cf  "}ul.lst-kix_x7t5b7jbczs4-4{list-style-type:none}ul.lst-kix_3kgcjgrfg4nh-6{list-style-type:none}ul.lst-kix_b24v03uyl4k9-1{list-style-type:none}ul.lst-kix_x7t5b7jbczs4-5{list-style-type:none}ul.lst-kix_3kgcjgrfg4nh-5{list-style-type:none}ul.lst-kix_b24v03uyl4k9-0{list-style-type:none}ul.lst-kix_x7t5b7jbczs4-6{list-style-type:none}ul.lst-kix_3kgcjgrfg4nh-8{list-style-type:none}.lst-kix_ao4z72tcnso6-7>li:before{content:"\0025cb  "}ul.lst-kix_x7t5b7jbczs4-7{list-style-type:none}ul.lst-kix_3kgcjgrfg4nh-7{list-style-type:none}ul.lst-kix_9a6zvrglxdu8-0{list-style-type:none}ul.lst-kix_x7t5b7jbczs4-0{list-style-type:none}ul.lst-kix_x7t5b7jbczs4-1{list-style-type:none}ul.lst-kix_x7t5b7jbczs4-2{list-style-type:none}ul.lst-kix_x7t5b7jbczs4-3{list-style-type:none}ul.lst-kix_9a6zvrglxdu8-4{list-style-type:none}ul.lst-kix_48dltnwvsqfp-2{list-style-type:none}ul.lst-kix_9a6zvrglxdu8-3{list-style-type:none}ul.lst-kix_48dltnwvsqfp-3{list-style-type:none}ul.lst-kix_9a6zvrglxdu8-2{list-style-type:none}ul.lst-kix_48dltnwvsqfp-0{list-style-type:none}ul.lst-kix_9a6zvrglxdu8-1{list-style-type:none}ul.lst-kix_48dltnwvsqfp-1{list-style-type:none}ul.lst-kix_48dltnwvsqfp-7{list-style-type:none}.lst-kix_euwmeg8cbpt8-1>li:before{content:"\0025cb  "}ul.lst-kix_48dltnwvsqfp-6{list-style-type:none}ul.lst-kix_48dltnwvsqfp-5{list-style-type:none}ul.lst-kix_48dltnwvsqfp-4{list-style-type:none}ul.lst-kix_48dltnwvsqfp-8{list-style-type:none}ul.lst-kix_3kgcjgrfg4nh-1{list-style-type:none}ul.lst-kix_b24v03uyl4k9-4{list-style-type:none}ul.lst-kix_3kgcjgrfg4nh-2{list-style-type:none}ul.lst-kix_b24v03uyl4k9-5{list-style-type:none}.lst-kix_velyrizglpvf-4>li:before{content:"\0025cb  "}.lst-kix_kmsnp6mmp404-6>li:before{content:"\0025cf  "}ul.lst-kix_3kgcjgrfg4nh-3{list-style-type:none}ul.lst-kix_b24v03uyl4k9-2{list-style-type:none}.lst-kix_riib1nssx348-3>li:before{content:"\0025cf  "}ul.lst-kix_3kgcjgrfg4nh-4{list-style-type:none}ul.lst-kix_b24v03uyl4k9-3{list-style-type:none}ul.lst-kix_b24v03uyl4k9-8{list-style-type:none}ul.lst-kix_b24v03uyl4k9-6{list-style-type:none}.lst-kix_kmsnp6mmp404-8>li:before{content:"\0025a0  "}ul.lst-kix_3kgcjgrfg4nh-0{list-style-type:none}ul.lst-kix_b24v03uyl4k9-7{list-style-type:none}.lst-kix_9drbe3jtkkuh-2>li{counter-increment:lst-ctn-kix_9drbe3jtkkuh-2}.lst-kix_l7t8lma3v5dd-4>li:before{content:"\0025cb  "}.lst-kix_4ed4y1s6lj5w-5>li:before{content:"\0025a0  "}ol.lst-kix_9drbe3jtkkuh-1.start{counter-reset:lst-ctn-kix_9drbe3jtkkuh-1 0}.lst-kix_xc244bhq3ube-5>li:before{content:"\0025a0  "}.lst-kix_tjc8wwk9cseo-6>li:before{content:"\0025cf  "}.lst-kix_sjinkw26u0wp-0>li:before{content:"\0025cf  "}.lst-kix_dpynbghv8b9j-4>li:before{content:"\0025cb  "}.lst-kix_riib1nssx348-0>li:before{content:"\0025cf  "}.lst-kix_pdbqztwjr65g-1>li:before{content:"\0025cb  "}.lst-kix_jamjx369pdr8-1>li:before{content:"\0025cb  "}.lst-kix_xc244bhq3ube-0>li:before{content:"\0025cf  "}ul.lst-kix_9a6zvrglxdu8-5{list-style-type:none}.lst-kix_3kgcjgrfg4nh-5>li:before{content:"\0025a0  "}ul.lst-kix_9a6zvrglxdu8-6{list-style-type:none}ul.lst-kix_9a6zvrglxdu8-7{list-style-type:none}ul.lst-kix_9a6zvrglxdu8-8{list-style-type:none}.lst-kix_jamjx369pdr8-6>li:before{content:"\0025cf  "}.lst-kix_hzkhz4hhlr8l-7>li:before{content:"\0025cb  "}.lst-kix_whzqgipsg1bt-0>li:before{content:"\0025cf  "}ol.lst-kix_9drbe3jtkkuh-8.start{counter-reset:lst-ctn-kix_9drbe3jtkkuh-8 0}.lst-kix_cbmcswhr1ftl-7>li:before{content:"\0025cb  "}.lst-kix_q40ki98g1i63-2>li:before{content:"\0025a0  "}.lst-kix_puzs83wo6j0o-6>li:before{content:"\0025cf  "}.lst-kix_myo05blafbk-1>li:before{content:"\0025cb  "}.lst-kix_jamjx369pdr8-0>li:before{content:"\0025cf  "}.lst-kix_yghjnj8s9xuc-3>li:before{content:"\0025cf  "}.lst-kix_riib1nssx348-6>li:before{content:"\0025cf  "}.lst-kix_yghjnj8s9xuc-5>li:before{content:"\0025a0  "}.lst-kix_7rljtqi4tpf8-4>li:before{content:"\0025cb  "}.lst-kix_cmqbbecr9wqr-1>li:before{content:"\0025cb  "}.lst-kix_puzs83wo6j0o-1>li:before{content:"\0025cb  "}.lst-kix_3jnvhqhv71nk-0>li:before{content:"\0025cf  "}ul.lst-kix_kmsnp6mmp404-0{list-style-type:none}ul.lst-kix_kmsnp6mmp404-1{list-style-type:none}.lst-kix_7rljtqi4tpf8-0>li:before{content:"\0025cf  "}.lst-kix_3kgcjgrfg4nh-6>li:before{content:"\0025cf  "}.lst-kix_vqtjz6k7exn7-6>li:before{content:"\0025cf  "}.lst-kix_whzqgipsg1bt-1>li:before{content:"\0025cb  "}ul.lst-kix_kmsnp6mmp404-4{list-style-type:none}ul.lst-kix_kmsnp6mmp404-5{list-style-type:none}ul.lst-kix_kmsnp6mmp404-2{list-style-type:none}ul.lst-kix_kmsnp6mmp404-3{list-style-type:none}.lst-kix_g3lepnw8n0x6-2>li:before{content:"\0025a0  "}.lst-kix_9a6zvrglxdu8-5>li:before{content:"\0025a0  "}ul.lst-kix_kmsnp6mmp404-8{list-style-type:none}ul.lst-kix_kmsnp6mmp404-7{list-style-type:none}.lst-kix_4ed4y1s6lj5w-3>li:before{content:"\0025cf  "}ul.lst-kix_kmsnp6mmp404-6{list-style-type:none}.lst-kix_10avetia2xt4-0>li:before{content:"\0025cf  "}.lst-kix_hzkhz4hhlr8l-2>li:before{content:"\0025a0  "}.lst-kix_g3lepnw8n0x6-1>li:before{content:"\0025cb  "}.lst-kix_kmsnp6mmp404-0>li:before{content:"\0025cf  "}.lst-kix_7rljtqi4tpf8-1>li:before{content:"\0025cb  "}ul.lst-kix_vqtjz6k7exn7-2{list-style-type:none}ul.lst-kix_10avetia2xt4-6{list-style-type:none}ul.lst-kix_vqtjz6k7exn7-3{list-style-type:none}ul.lst-kix_10avetia2xt4-7{list-style-type:none}ul.lst-kix_vqtjz6k7exn7-0{list-style-type:none}ul.lst-kix_10avetia2xt4-4{list-style-type:none}.lst-kix_cmqbbecr9wqr-7>li:before{content:"\0025cb  "}ul.lst-kix_vqtjz6k7exn7-1{list-style-type:none}ul.lst-kix_10avetia2xt4-5{list-style-type:none}ul.lst-kix_10avetia2xt4-2{list-style-type:none}ul.lst-kix_10avetia2xt4-3{list-style-type:none}.lst-kix_fscfg01f3awm-5>li:before{content:"\0025a0  "}.lst-kix_vqtjz6k7exn7-2>li:before{content:"\0025a0  "}ul.lst-kix_10avetia2xt4-0{list-style-type:none}.lst-kix_pdbqztwjr65g-2>li:before{content:"\0025a0  "}ul.lst-kix_10avetia2xt4-1{list-style-type:none}ul.lst-kix_vqtjz6k7exn7-8{list-style-type:none}ul.lst-kix_vqtjz6k7exn7-6{list-style-type:none}ul.lst-kix_vqtjz6k7exn7-7{list-style-type:none}.lst-kix_jamjx369pdr8-8>li:before{content:"\0025a0  "}ul.lst-kix_vqtjz6k7exn7-4{list-style-type:none}ul.lst-kix_10avetia2xt4-8{list-style-type:none}ul.lst-kix_vqtjz6k7exn7-5{list-style-type:none}.lst-kix_4ed4y1s6lj5w-4>li:before{content:"\0025cb  "}.lst-kix_pdbqztwjr65g-3>li:before{content:"\0025cf  "}.lst-kix_508olx789l51-7>li:before{content:"\0025cb  "}.lst-kix_gr4a4jnxityy-4>li:before{content:"\0025cb  "}.lst-kix_qzhxzy2sbaso-3>li:before{content:"\0025cf  "}.lst-kix_tjc8wwk9cseo-7>li:before{content:"\0025cb  "}ul.lst-kix_velyrizglpvf-0{list-style-type:none}.lst-kix_9a6zvrglxdu8-7>li:before{content:"\0025cb  "}ul.lst-kix_velyrizglpvf-2{list-style-type:none}ul.lst-kix_velyrizglpvf-1{list-style-type:none}ul.lst-kix_velyrizglpvf-4{list-style-type:none}ul.lst-kix_velyrizglpvf-3{list-style-type:none}ul.lst-kix_velyrizglpvf-6{list-style-type:none}ul.lst-kix_velyrizglpvf-5{list-style-type:none}ul.lst-kix_velyrizglpvf-8{list-style-type:none}ul.lst-kix_velyrizglpvf-7{list-style-type:none}.lst-kix_fscfg01f3awm-7>li:before{content:"\0025cb  "}.lst-kix_z0ttp4mliy4e-1>li:before{content:"\0025cb  "}.lst-kix_e5f9ugvogyec-4>li:before{content:"\0025cb  "}.lst-kix_vqtjz6k7exn7-4>li:before{content:"\0025cb  "}.lst-kix_cmqbbecr9wqr-5>li:before{content:"\0025a0  "}.lst-kix_7kppodwmns3b-7>li:before{content:"\0025cb  "}.lst-kix_velyrizglpvf-8>li:before{content:"\0025a0  "}.lst-kix_sjinkw26u0wp-8>li:before{content:"\0025a0  "}.lst-kix_x7t5b7jbczs4-3>li:before{content:"\0025cf  "}.lst-kix_cbmcswhr1ftl-8>li:before{content:"\0025a0  "}.lst-kix_riib1nssx348-1>li:before{content:"\0025cb  "}.lst-kix_myo05blafbk-2>li:before{content:"\0025a0  "}.lst-kix_4tq1geabcxa-0>li:before{content:"\0025cf  "}ul.lst-kix_pdbqztwjr65g-2{list-style-type:none}ul.lst-kix_pdbqztwjr65g-1{list-style-type:none}ul.lst-kix_pdbqztwjr65g-0{list-style-type:none}.lst-kix_g96zcqu3kh41-2>li:before{content:"\0025a0  "}ul.lst-kix_pdbqztwjr65g-6{list-style-type:none}ul.lst-kix_pdbqztwjr65g-5{list-style-type:none}ul.lst-kix_pdbqztwjr65g-4{list-style-type:none}ul.lst-kix_pdbqztwjr65g-3{list-style-type:none}ul.lst-kix_pdbqztwjr65g-8{list-style-type:none}ul.lst-kix_pdbqztwjr65g-7{list-style-type:none}.lst-kix_dpynbghv8b9j-1>li:before{content:"\0025cb  "}.lst-kix_3jnvhqhv71nk-1>li:before{content:"\0025cb  "}.lst-kix_3kgcjgrfg4nh-0>li:before{content:"\0025cf  "}.lst-kix_cbmcswhr1ftl-3>li:before{content:"\0025cf  "}.lst-kix_ao4z72tcnso6-5>li:before{content:"\0025a0  "}.lst-kix_cbmcswhr1ftl-4>li:before{content:"\0025cb  "}.lst-kix_hzkhz4hhlr8l-5>li:before{content:"\0025a0  "}.lst-kix_9drbe3jtkkuh-0>li:before{content:"" counter(lst-ctn-kix_9drbe3jtkkuh-0,decimal) ". "}ul.lst-kix_508olx789l51-2{list-style-type:none}ul.lst-kix_508olx789l51-1{list-style-type:none}ul.lst-kix_508olx789l51-0{list-style-type:none}.lst-kix_4tq1geabcxa-3>li:before{content:"\0025cf  "}.lst-kix_kmsnp6mmp404-2>li:before{content:"\0025a0  "}.lst-kix_508olx789l51-2>li:before{content:"\0025a0  "}ul.lst-kix_508olx789l51-7{list-style-type:none}ul.lst-kix_508olx789l51-8{list-style-type:none}ul.lst-kix_508olx789l51-5{list-style-type:none}ul.lst-kix_508olx789l51-6{list-style-type:none}ul.lst-kix_508olx789l51-3{list-style-type:none}ul.lst-kix_508olx789l51-4{list-style-type:none}.lst-kix_g3lepnw8n0x6-7>li:before{content:"\0025cb  "}.lst-kix_euwmeg8cbpt8-5>li:before{content:"\0025a0  "}.lst-kix_508olx789l51-8>li:before{content:"\0025a0  "}.lst-kix_g96zcqu3kh41-4>li:before{content:"\0025cb  "}ul.lst-kix_tjc8wwk9cseo-1{list-style-type:none}.lst-kix_9drbe3jtkkuh-7>li:before{content:"" counter(lst-ctn-kix_9drbe3jtkkuh-7,lower-latin) ". "}ul.lst-kix_tjc8wwk9cseo-0{list-style-type:none}ul.lst-kix_owpom8h3nkwo-8{list-style-type:none}ul.lst-kix_tjc8wwk9cseo-3{list-style-type:none}ul.lst-kix_owpom8h3nkwo-7{list-style-type:none}ul.lst-kix_tjc8wwk9cseo-2{list-style-type:none}ul.lst-kix_tjc8wwk9cseo-5{list-style-type:none}ul.lst-kix_tjc8wwk9cseo-4{list-style-type:none}ul.lst-kix_owpom8h3nkwo-4{list-style-type:none}ul.lst-kix_tjc8wwk9cseo-7{list-style-type:none}ul.lst-kix_owpom8h3nkwo-3{list-style-type:none}ul.lst-kix_tjc8wwk9cseo-6{list-style-type:none}ul.lst-kix_owpom8h3nkwo-6{list-style-type:none}ul.lst-kix_owpom8h3nkwo-5{list-style-type:none}ul.lst-kix_tjc8wwk9cseo-8{list-style-type:none}ul.lst-kix_owpom8h3nkwo-0{list-style-type:none}ul.lst-kix_owpom8h3nkwo-2{list-style-type:none}ul.lst-kix_owpom8h3nkwo-1{list-style-type:none}.lst-kix_riib1nssx348-4>li:before{content:"\0025cb  "}.lst-kix_yghjnj8s9xuc-8>li:before{content:"\0025a0  "}.lst-kix_l7t8lma3v5dd-8>li:before{content:"\0025a0  "}.lst-kix_dpynbghv8b9j-8>li:before{content:"\0025a0  "}.lst-kix_riib1nssx348-7>li:before{content:"\0025cb  "}.lst-kix_3kgcjgrfg4nh-8>li:before{content:"\0025a0  "}.lst-kix_g96zcqu3kh41-5>li:before{content:"\0025a0  "}.lst-kix_qzhxzy2sbaso-5>li:before{content:"\0025a0  "}.lst-kix_4tq1geabcxa-8>li:before{content:"\0025a0  "}.lst-kix_3jnvhqhv71nk-8>li:before{content:"\0025a0  "}.lst-kix_9drbe3jtkkuh-3>li{counter-increment:lst-ctn-kix_9drbe3jtkkuh-3}.lst-kix_velyrizglpvf-3>li:before{content:"\0025cf  "}.lst-kix_riib1nssx348-2>li:before{content:"\0025a0  "}.lst-kix_b24v03uyl4k9-3>li:before{content:"\0025cf  "}.lst-kix_dpynbghv8b9j-2>li:before{content:"\0025a0  "}.lst-kix_g3lepnw8n0x6-3>li:before{content:"\0025cf  "}ul.lst-kix_4ed4y1s6lj5w-0{list-style-type:none}ul.lst-kix_4ed4y1s6lj5w-4{list-style-type:none}ul.lst-kix_4ed4y1s6lj5w-3{list-style-type:none}ul.lst-kix_4ed4y1s6lj5w-2{list-style-type:none}ul.lst-kix_4ed4y1s6lj5w-1{list-style-type:none}.lst-kix_gr4a4jnxityy-5>li:before{content:"\0025a0  "}ul.lst-kix_3jnvhqhv71nk-2{list-style-type:none}.lst-kix_olh7ea290trr-1>li:before{content:"\0025cb  "}ul.lst-kix_3jnvhqhv71nk-3{list-style-type:none}ul.lst-kix_3jnvhqhv71nk-0{list-style-type:none}ul.lst-kix_3jnvhqhv71nk-1{list-style-type:none}ul.lst-kix_cbmcswhr1ftl-0{list-style-type:none}.lst-kix_x7iifx7mthae-4>li:before{content:"\0025cb  "}ul.lst-kix_cbmcswhr1ftl-6{list-style-type:none}ul.lst-kix_cbmcswhr1ftl-5{list-style-type:none}ul.lst-kix_cbmcswhr1ftl-8{list-style-type:none}ul.lst-kix_3jnvhqhv71nk-8{list-style-type:none}ul.lst-kix_cbmcswhr1ftl-7{list-style-type:none}ul.lst-kix_3jnvhqhv71nk-7{list-style-type:none}ul.lst-kix_cbmcswhr1ftl-2{list-style-type:none}ul.lst-kix_3jnvhqhv71nk-6{list-style-type:none}ul.lst-kix_cbmcswhr1ftl-1{list-style-type:none}ul.lst-kix_3jnvhqhv71nk-5{list-style-type:none}ul.lst-kix_cbmcswhr1ftl-4{list-style-type:none}ul.lst-kix_3jnvhqhv71nk-4{list-style-type:none}ul.lst-kix_cbmcswhr1ftl-3{list-style-type:none}.lst-kix_9drbe3jtkkuh-2>li:before{content:"" counter(lst-ctn-kix_9drbe3jtkkuh-2,lower-roman) ". "}ul.lst-kix_7kppodwmns3b-0{list-style-type:none}ul.lst-kix_7kppodwmns3b-1{list-style-type:none}ul.lst-kix_7kppodwmns3b-4{list-style-type:none}.lst-kix_e5f9ugvogyec-2>li:before{content:"\0025a0  "}.lst-kix_myo05blafbk-8>li:before{content:"\0025a0  "}ul.lst-kix_7kppodwmns3b-5{list-style-type:none}ul.lst-kix_7kppodwmns3b-2{list-style-type:none}.lst-kix_cmqbbecr9wqr-4>li:before{content:"\0025cb  "}ul.lst-kix_7kppodwmns3b-3{list-style-type:none}.lst-kix_olh7ea290trr-5>li:before{content:"\0025a0  "}ul.lst-kix_7kppodwmns3b-8{list-style-type:none}ul.lst-kix_7kppodwmns3b-6{list-style-type:none}ul.lst-kix_7kppodwmns3b-7{list-style-type:none}.lst-kix_9drbe3jtkkuh-5>li{counter-increment:lst-ctn-kix_9drbe3jtkkuh-5}.lst-kix_olh7ea290trr-2>li:before{content:"\0025a0  "}.lst-kix_l7t8lma3v5dd-0>li:before{content:"\0025cf  "}.lst-kix_whzqgipsg1bt-4>li:before{content:"\0025cb  "}.lst-kix_cmqbbecr9wqr-3>li:before{content:"\0025cf  "}.lst-kix_z0ttp4mliy4e-5>li:before{content:"\0025a0  "}.lst-kix_vqtjz6k7exn7-8>li:before{content:"\0025a0  "}ul.lst-kix_xc244bhq3ube-2{list-style-type:none}ul.lst-kix_xc244bhq3ube-3{list-style-type:none}ul.lst-kix_xc244bhq3ube-0{list-style-type:none}ul.lst-kix_xc244bhq3ube-1{list-style-type:none}ul.lst-kix_xc244bhq3ube-6{list-style-type:none}ul.lst-kix_xc244bhq3ube-7{list-style-type:none}ul.lst-kix_xc244bhq3ube-4{list-style-type:none}ul.lst-kix_xc244bhq3ube-5{list-style-type:none}.lst-kix_3kgcjgrfg4nh-4>li:before{content:"\0025cb  "}ul.lst-kix_xc244bhq3ube-8{list-style-type:none}.lst-kix_l7t8lma3v5dd-6>li:before{content:"\0025cf  "}.lst-kix_7rljtqi4tpf8-7>li:before{content:"\0025cb  "}.lst-kix_x7iifx7mthae-2>li:before{content:"\0025a0  "}.lst-kix_x7iifx7mthae-3>li:before{content:"\0025cf  "}.lst-kix_10avetia2xt4-5>li:before{content:"\0025a0  "}.lst-kix_pdbqztwjr65g-0>li:before{content:"\0025cf  "}ul.lst-kix_x7iifx7mthae-8{list-style-type:none}ul.lst-kix_x7iifx7mthae-6{list-style-type:none}ul.lst-kix_x7iifx7mthae-7{list-style-type:none}ol.lst-kix_9drbe3jtkkuh-7.start{counter-reset:lst-ctn-kix_9drbe3jtkkuh-7 0}ul.lst-kix_x7iifx7mthae-4{list-style-type:none}ul.lst-kix_x7iifx7mthae-5{list-style-type:none}ul.lst-kix_x7iifx7mthae-2{list-style-type:none}ul.lst-kix_x7iifx7mthae-3{list-style-type:none}.lst-kix_owpom8h3nkwo-5>li:before{content:"\0025a0  "}.lst-kix_g3lepnw8n0x6-8>li:before{content:"\0025a0  "}.lst-kix_owpom8h3nkwo-0>li:before{content:"\0025cf  "}ul.lst-kix_x7iifx7mthae-1{list-style-type:none}ul.lst-kix_x7iifx7mthae-0{list-style-type:none}.lst-kix_velyrizglpvf-2>li:before{content:"\0025a0  "}.lst-kix_vz631vhbp5pc-7>li:before{content:"\0025cb  "}.lst-kix_e5f9ugvogyec-5>li:before{content:"\0025a0  "}.lst-kix_olh7ea290trr-8>li:before{content:"\0025a0  "}.lst-kix_olh7ea290trr-3>li:before{content:"\0025cf  "}.lst-kix_g3lepnw8n0x6-6>li:before{content:"\0025cf  "}.lst-kix_cmqbbecr9wqr-8>li:before{content:"\0025a0  "}.lst-kix_z0ttp4mliy4e-4>li:before{content:"\0025cb  "}.lst-kix_4tq1geabcxa-6>li:before{content:"\0025cf  "}ul.lst-kix_euwmeg8cbpt8-2{list-style-type:none}ul.lst-kix_euwmeg8cbpt8-3{list-style-type:none}ul.lst-kix_euwmeg8cbpt8-0{list-style-type:none}ul.lst-kix_euwmeg8cbpt8-1{list-style-type:none}.lst-kix_qzhxzy2sbaso-8>li:before{content:"\0025a0  "}ul.lst-kix_euwmeg8cbpt8-6{list-style-type:none}.lst-kix_xc244bhq3ube-6>li:before{content:"\0025cf  "}ul.lst-kix_euwmeg8cbpt8-7{list-style-type:none}ul.lst-kix_euwmeg8cbpt8-4{list-style-type:none}.lst-kix_hzkhz4hhlr8l-8>li:before{content:"\0025a0  "}ul.lst-kix_euwmeg8cbpt8-5{list-style-type:none}ul.lst-kix_euwmeg8cbpt8-8{list-style-type:none}ol{margin:0;padding:0}.c18{vertical-align:top;width:468pt;border-style:solid;background-color:#b6d7a8;border-color:#000000;border-width:1pt;padding:5pt 5pt 5pt 5pt}.c4{vertical-align:top;width:468pt;border-style:solid;background-color:#f9cb9c;border-color:#000000;border-width:1pt;padding:5pt 5pt 5pt 5pt}.c19{vertical-align:top;width:51.8pt;border-style:solid;border-color:#000000;border-width:1pt;padding:5pt 5pt 5pt 5pt}.c24{vertical-align:top;width:60pt;border-style:solid;border-color:#000000;border-width:1pt;padding:5pt 5pt 5pt 5pt}.c28{vertical-align:top;width:61.5pt;border-style:solid;border-color:#000000;border-width:1pt;padding:5pt 5pt 5pt 5pt}.c16{vertical-align:top;width:157.5pt;border-style:solid;border-color:#000000;border-width:1pt;padding:5pt 5pt 5pt 5pt}.c23{vertical-align:top;width:81pt;border-style:solid;border-color:#000000;border-width:1pt;padding:5pt 5pt 5pt 5pt}.c20{vertical-align:top;width:56.2pt;border-style:solid;border-color:#000000;border-width:1pt;padding:5pt 5pt 5pt 5pt}.c8{color:#cc0000;font-size:14pt;font-weight:bold}.c30{margin-right:auto;border-collapse:collapse}.c1{padding-left:0pt;line-height:1.0;margin-left:36pt}.c3{max-width:468pt;background-color:#ffffff;padding:72pt 72pt 72pt 72pt}.c31{color:inherit;text-decoration:inherit}.c32{padding-left:4.5pt;margin-left:30.8pt}.c22{padding-left:0pt;margin-left:72pt}.c34{text-indent:-18pt;margin-left:36pt}.c5{font-size:10pt;font-family:"Courier New"}.c35{padding-top:14pt;padding-bottom:14pt}.c14{margin:0;padding:0}.c7{padding-left:0pt;margin-left:36pt}.c12{text-decoration:underline}.c11{font-weight:bold}.c29{font-style:italic}.c17{height:0pt}.c10{text-align:justify}.c33{color:#ff0000}.c27{text-indent:36pt}.c15{font-size:12pt}.c2{line-height:1.0}.c25{color:#cc0000}.c26{color:#1155cc}.c9{font-family:"Courier New"}.c13{text-align:center}.c0{direction:ltr}.c21{font-size:14pt}.c6{height:11pt}.title{padding-top:0pt;line-height:1.15;text-align:left;color:#000000;font-size:21pt;font-family:"Trebuchet MS";padding-bottom:0pt}.subtitle{padding-top:0pt;line-height:1.15;text-align:left;color:#666666;font-style:italic;font-size:13pt;font-family:"Trebuchet MS";padding-bottom:10pt}li{color:#000000;font-size:11pt;font-family:"Arial"}p{color:#000000;font-size:11pt;margin:0;font-family:"Arial"}h1{padding-top:10pt;line-height:1.15;text-align:left;color:#000000;font-size:16pt;font-family:"Trebuchet MS";padding-bottom:0pt}h2{padding-top:10pt;line-height:1.15;text-align:left;color:#000000;font-size:13pt;font-family:"Trebuchet MS";font-weight:bold;padding-bottom:0pt}h3{padding-top:8pt;line-height:1.15;text-align:left;color:#666666;font-size:12pt;font-family:"Trebuchet MS";font-weight:bold;padding-bottom:0pt}h4{padding-top:8pt;line-height:1.15;text-align:left;color:#666666;font-size:11pt;text-decoration:underline;font-family:"Trebuchet MS";padding-bottom:0pt}h5{padding-top:8pt;line-height:1.15;text-align:left;color:#666666;font-size:11pt;font-family:"Trebuchet MS";padding-bottom:0pt}h6{padding-top:8pt;line-height:1.15;text-align:left;color:#666666;font-style:italic;font-size:11pt;font-family:"Trebuchet MS";padding-bottom:0pt}</style></head><body class="c3"><p class="c0"><span class="c8 c12">Sections</span></p><p class="c0"><span class="c8">A. Why you may need this</span></p><p class="c0"><span class="c8">B. Validations</span></p><p class="c0"><span class="c8">C. What&rsquo;s the structure of the code</span></p><p class="c0"><span class="c8">D. How to run the tool</span></p><p class="c0"><span class="c8">E. Write a new validation</span></p><p class="c0"><span class="c8">F. Output results</span></p><p class="c13 c0 c6"><span class="c8"></span></p><p class="c13 c0 c6"><span class="c8"></span></p><p class="c13 c0"><span class="c8">A. Why you may need this</span></p><p class="c0 c6"><span class="c11"></span></p><p class="c10 c0"><span class="c11">The amount of data that has being generated in laboratories exceeds the human resources that can check and validate this data by hand. See example below: swiss-prot contains human-checked proteins, while TrEmbl contains all the proteins sequenced that are waiting to be validated and to be assigned a biological function.</span></p><p class="c0"><span class="c11">&nbsp;</span></p><p class="c13 c0"><img height="431" src="images/image16.jpg" width="560"></p><p class="c0"><span>&copy; http://gorbi.irb.hr/en/method/growth-of-sequence-databases/</span><hr style="page-break-before:always;display:none;"></p><p class="c13 c0"><span class="c8">B. Validations</span></p><p class="c13 c0 c6"><span></span></p><p class="c10 c0"><span class="c11">We designed a tool that helps identifying problems with gene predictions, based on similarities with data from public databases (like Swissprot and Uniprot). We apply a set of validation tests that provide useful information about the problems that appear in the predictions and how they can be curated. Our target users are the biologists, who have big amounts of genetic data that has to be manually curated, and our tool will prioritize the genes that need to be checked and suggest possible error causes.</span></p><p class="c10 c0 c6"><span class="c11"></span></p><p class="c0"><span class="c11">1) Length validation via clusterization</span></p><p class="c0"><span class="c11">2) Length validation via ranking</span></p><p class="c0"><span class="c11">3) Duplication</span></p><p class="c0"><span class="c11">4) Gene merge</span></p><p class="c0"><span class="c11">5) Multiple alignment</span></p><p class="c0"><span class="c11">6) Blast Reading Frame - applicable only for nucleotide sequences</span></p><p class="c0"><span class="c11">7) Open reading frame - applicable only for nucleotide sequences</span></p><p class="c0"><span class="c11 c29">8) Codon usage -- under development</span></p><p class="c0 c6"><span class="c11"></span></p><p class="c0 c6"><span class="c11"></span></p><p class="c0"><span class="c11">Data needed for each validation</span><img height="224" src="images/image14.jpg" width="781"></p><p class="c0 c6"><span class="c11"></span></p><p class="c0 c6"><span class="c11"></span></p><p class="c0 c6"><span class="c11"></span></p><a href="#" name="fb9fb93011ad78ca1e36c4c1e56fdf04d009bd67"></a><a href="#" name="0"></a><table cellpadding="0" cellspacing="0" class="c30"><tbody></tbody></table><p class="c0 c6"><span class="c11"></span></p><p class="c0 c6"><span class="c11"></span></p><p class="c0 c6"><span class="c11"></span></p><hr style="page-break-before:always;display:none;"><p class="c0 c6"><span class="c11"></span></p><p class="c0"><span class="c11">1) Length Validation via clusterization</span></p><p class="c0 c6"><span></span></p><a href="#" name="0e18eeb8ff26ff9057465abf0461155a56a82e14"></a><a href="#" name="1"></a><table cellpadding="0" cellspacing="0" class="c30"><tbody><tr class="c17"><td class="c18"><p class="c2 c0"><span class="c11">Error causes</span></p><ul class="c14 lst-kix_3jnvhqhv71nk-0 start"><li class="c1 c0"><span>sequencing error: some parts of the gene were lost/added on the way</span></li></ul><ul class="c14 lst-kix_mk8t5sy0k589-0 start"><li class="c0 c1"><span>gene prediction error: after &ldquo;reading&rdquo; a genome sequence with new generation sequencing techniques, the start/end of the gene was not well estimated</span></li><li class="c1 c0"><span>the gene had a low expression level</span></li><li class="c1 c0"><span>the sequenced mrna incorrectly contains an some introns </span></li></ul><p class="c2 c0 c6"><span></span></p></td></tr><tr><td class="c18"><p class="c2 c0"><span class="c11">Input data</span></p><ul class="c14 lst-kix_4ed4y1s6lj5w-0 start"><li class="c7 c0"><span>lengths of the hits: this data is retrieved after parsing the blast output file</span></li><li class="c0 c7"><span>length of the prediction: this number is the length of the current query from the fasta. In case of nucleotide sequences, we are interested in the length of the corresponding protein translation of the query (which is the length of the query divided by 3, plus or minus 1 or 2 residues, depending on the reading frame -- this detail is not taken into consideration here)</span></li></ul><p class="c2 c0 c6"><span></span></p></td></tr><tr class="c17"><td class="c18"><p class="c2 c0"><span class="c11">Class information:</span></p><ul class="c14 lst-kix_7rljtqi4tpf8-0 start"><li class="c1 c0"><span>short_header: LengthCluster</span></li><li class="c1 c0"><span>header: Length Cluster</span></li><li class="c1 c0"><span>short description: Check whether the prediction length fits most of the BLAST hit lengths, by 1D hierarchical clusterization. Meaning of the output displayed: Prediction_len [Main Cluster Length Interval]</span></li><li class="c1 c0"><span>cli_name: lenc</span></li></ul></td></tr><tr><td class="c18"><p class="c2 c0"><span class="c11">Workflow</span></p><p class="c2 c0"><span class="c11">Aim :</span><span>&nbsp;we are interested to find out if the length of the predicted sequence belongs to the distribution of the hit lengths (in other words, how close is the length of the prediction to the majority of the lengths of the hits)</span></p><p class="c2 c0 c6"><span></span></p><p class="c10 c2 c0"><span class="c11">(1)</span><span>&nbsp;By plotting the histogram of the length distribution of the hits we observe that the distribution does not fit a Bell Curve, so we cannot apply the classical T-test. </span></p><p class="c10 c2 c0 c6"><span></span></p><p class="c10 c2 c0 c6"><span></span></p><p class="c10 c2 c0 c6"><span></span></p><p class="c2 c13 c0"><img height="310" src="images/image09.jpg" width="310"></p><p class="c2 c13 c0"><span class="c11">(1)</span></p><p class="c10 c2 c0 c6"><span></span></p><p class="c10 c2 c0"><span>Our approach to find the majority of lengths among the reference lengths &nbsp;is by a typical </span><span class="c11">hierarchical clusterization</span><span>: </span></p><p class="c10 c2 c0 c6"><span></span></p><p class="c10 c2 c0"><span class="c11">(2)</span><span>&nbsp;Firstly we assume that each length belongs to a separate cluster. Each step we merge the closest two clusters, until a cluster that contains more than 50% of the reference sequences is obtained. It is called the &ldquo;main cluster&rdquo; (the one colored in red on the histogram).</span></p><p class="c10 c2 c0 c6"><span></span></p><p class="c10 c2 c0 c6"><span></span></p><p class="c2 c13 c0"><img height="308" src="images/image07.jpg" width="308"></p><p class="c2 c13 c0"><span class="c11">(2)</span></p><p class="c2 c0 c6 c13"><span></span></p><p class="c10 c2 c0"><span class="c11">(3) </span><span>Finally we are interested to check whether the length of the prediction belongs to the main cluster or not and to return a message accordingly. In Figure 3 the prediction belongs to the main cluster and the query passes this validation test. &nbsp; </span></p><p class="c10 c2 c0 c6"><span></span></p><p class="c10 c2 c0 c6"><span></span></p><p class="c2 c13 c0"><img height="298" src="images/image08.jpg" width="310"></p><p class="c2 c13 c0"><span class="c11">(3)</span></p><p class="c2 c0 c6"><span class="c11"></span></p></td></tr><tr><td class="c18"><p class="c2 c0"><span class="c11">Code [1]</span></p><p class="c2 c0 c6"><span class="c11"></span></p><p class="c2 c0"><span class="c5">#clusterization by length</span></p><p class="c2 c0"><span class="c5">contents = lst.map{ |x| x.length_protein.to_i }.sort{|a,b| a&lt;=&gt;b}</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5">hc = HierarchicalClusterization.new(contents)</span></p><p class="c2 c0"><span class="c5">clusters = hc.hierarchical_clusterization</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5">max_density = 0;</span></p><p class="c2 c0"><span class="c5">max_density_cluster_idx = 0;</span></p><p class="c2 c0"><span class="c5">clusters.each_with_index do |item, i|</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;if item.density &gt; max_density</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;max_density = item.density</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;max_density_cluster_idx = i;</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;end</span></p><p class="c2 c0"><span class="c5">end</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5">clusterization = [clusters, max_density_cluster_idx]</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5">@clusters = clusterization[0]</span></p><p class="c2 c0"><span class="c5">@max_density_cluster = clusterization[1]</span></p><p class="c2 c0"><span class="c5">limits = @clusters[@max_density_cluster].get_limits</span></p><p class="c2 c0"><span class="c5">prediction_len = @prediction.length_protein</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5">@validation_report = LengthClusterValidationOutput.new(prediction_len, limits)</span></p><p class="c2 c0"><span class="c5">plot1 = plot_histo_clusters</span></p><p class="c2 c0"><span class="c5">@validation_report.plot_files.push(plot1)</span></p><p class="c2 c0"><span class="c5">plot2 = plot_len_clusters</span></p><p class="c2 c0"><span class="c5">@validation_report.plot_files.push(plot2)</span></p><p class="c2 c0"><span class="c5">[1] </span><span class="c5 c12 c26"><a class="c31" href="https://github.com/monicadragan/GeneValidator/blob/rubygem/lib/genevalidator/validation_length_cluster.rb#L67">https://github.com/monicadragan/GeneValidator/blob/rubygem/lib/genevalidator/validation_length_cluster.rb#L67</a></span><span class="c5">&nbsp;</span></p></td></tr><tr><td class="c18"><p class="c2 c0"><span class="c11">Plots</span></p><ul class="c14 lst-kix_euwmeg8cbpt8-0 start"><li class="c1 c10 c0"><span>length distribution histogram (prediction passes the validation test in Figure 3 and fails in Figure 4)</span></li></ul><p class="c10 c2 c0 c6"><span></span></p><p class="c10 c2 c0 c6"><span></span></p><p class="c2 c13 c0"><img height="205" src="images/image08.jpg" width="213"><img height="204" src="images/image02.jpg" width="216"></p><p class="c2 c13 c0"><span class="c11">&nbsp; (3) &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; (4) </span></p><p class="c2 c13 c0 c6"><span class="c11"></span></p><ul class="c14 lst-kix_x7iifx7mthae-0 start"><li class="c1 c0"><span>line plot representing the regions of the hits that are matched by the prediction query</span></li></ul><p class="c2 c0 c6"><span></span></p><p class="c2 c13 c0"><img height="315" src="images/image00.png" width="381"></p><p class="c2 c0 c6"><span class="c11"></span></p></td></tr></tbody></table><p class="c0 c6"><span></span></p><p class="c0"><span class="c11">2) Length Validation via ranking</span></p><p class="c0 c6"><span></span></p><a href="#" name="c8cb6ea51b906c1d260b1569bf10c5e57ac64a94"></a><a href="#" name="2"></a><table cellpadding="0" cellspacing="0" class="c30"><tbody><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Error causes</span></p><ul class="c14 lst-kix_mk8t5sy0k589-0"><li class="c1 c0"><span>sequencing error: some parts of the gene were lost/added on the way</span></li><li class="c1 c0"><span>gene prediction error: after &ldquo;reading&rdquo; a genome sequence with new generation sequencing techniques, the start/end of the gene was not well estimated</span></li><li class="c1 c0"><span>the gene had a low expression level</span></li><li class="c1 c0"><span>the introns were not well removed from the gene</span></li></ul></td></tr><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Input data</span></p><ul class="c14 lst-kix_4ed4y1s6lj5w-0"><li class="c7 c0"><span>lengths of the hits: this data is retrieved after parsing the blast output file</span></li><li class="c7 c0"><span>length of the prediction: this number is the length of the current query from the fasta. In case of nucleotide sequences, we are interested in the length of the corresponding protein translation of the query (which is the length of the query divided by 3, plus or minus 1 or 2 residues, depending on the reading frame -- this detail is not taken into consideration here)</span></li></ul></td></tr><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Class information:</span></p><ul class="c14 lst-kix_7kppodwmns3b-0 start"><li class="c1 c0"><span>short_header: LengthRank</span></li><li class="c1 c0"><span>header: Length Rank</span></li><li class="c1 c0"><span>description: Check whether the rank of the prediction length lies among 80% of all the BLAST hit lengths. Meaning of the output displayed: no of extreme length hits / total no of hits</span></li><li class="c1 c0"><span>cli_name: lenr</span></li></ul></td></tr><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Workflow</span></p><p class="c2 c0 c6"><span class="c11"></span></p><p class="c2 c0"><span class="c11">Aim: </span><span>We want to see how extreme the length of the prediction is among all the other hits. </span></p><p class="c2 c0"><span class="c11">Approach:</span></p><ul class="c14 lst-kix_velyrizglpvf-0 start"><li class="c1 c0"><span>sort the hit lengths in ascending order.</span></li><li class="c1 c0"><span>find the median for the set of hit lengths</span></li><li class="c1 c0"><span>find on which side of the median the length of the prediction belongs to</span></li><li class="c1 c0"><span>if the prediction length is on the right side of the media, count how many hit lengths are longer than the prediction length (Figure 1)</span></li><li class="c1 c0"><span>otherwise count how many hit lengths are shorter than the prediction length (Figure 2)</span></li></ul><p class="c2 c0 c6"><span></span></p><p class="c2 c13 c0"><img height="233" src="images/image13.png" width="247"><span>&nbsp; &nbsp; &nbsp; &nbsp;</span><img height="233" src="images/image15.png" width="240"></p><p class="c2 c13 c0"><span class="c11">Figure 1 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; Figure 2</span></p></td></tr><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Code [1]</span></p><p class="c2 c0 c6"><span class="c11"></span></p><p class="c2 c0"><span class="c5">threshold = 0.2</span></p><p class="c2 c0"><span class="c5">lengths = hits.map{ |x| x.length_protein.to_i }.sort{|a,b| a&lt;=&gt;b}</span></p><p class="c2 c0"><span class="c5">len = lengths.length</span></p><p class="c2 c0"><span class="c5">median = len % 2 == 1 ?</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;lengths[len/2] :</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;(lengths[len/2 - 1] + lengths[len/2]).to_f / 2</span></p><p class="c2 c0"><span class="c5">predicted_len = prediction.length_protein</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5">if hits.length == 1</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp;msg = &quot;&quot;</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp;percentage = 1</span></p><p class="c2 c0"><span class="c5">else</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp;if predicted_len &lt; median</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;rank = lengths.find_all{|x| x &lt; predicted_len}.length</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;percentage = rank / (len + 0.0)</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;msg = &quot;TOO_SHORT&quot;</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp;else</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;rank = lengths.find_all{|x| x &gt; predicted_len}.length</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;percentage = rank / (len + 0.0)</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;msg = &quot;TOO_LONG&quot;</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp;end</span></p><p class="c2 c0"><span class="c5">end</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5">if percentage &gt;= threshold</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;msg = &quot;&quot;</span></p><p class="c2 c0"><span class="c5">end</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;@validation_report = LengthRankValidationOutput.new(msg, percentage.round(2))</span></p><p class="c2 c0"><span class="c5">[1] </span><span class="c5 c12 c26"><a class="c31" href="https://github.com/monicadragan/GeneValidator/blob/rubygem/lib/genevalidator/validation_length_rank.rb#L65">https://github.com/monicadragan/GeneValidator/blob/rubygem/lib/genevalidator/validation_length_rank.rb#L65</a></span><span class="c5">&nbsp;</span></p><p class="c2 c0 c6"><span class="c5"></span></p></td></tr><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Plots</span></p><p class="c2 c0"><span>No plots are generated</span></p></td></tr></tbody></table><p class="c0"><span class="c11">3) &nbsp;Duplication Check</span></p><p class="c0 c6"><span></span></p><a href="#" name="91de52647cf180b5783e48991e5d11180e91820a"></a><a href="#" name="3"></a><table cellpadding="0" cellspacing="0" class="c30"><tbody><tr class="c17"><td class="c18"><p class="c2 c0"><span class="c11">Error causes</span></p><ul class="c14 lst-kix_9a6zvrglxdu8-0 start"><li class="c1 c0"><span>sequencing error</span></li><li class="c1 c0"><span>error in the replication mechanism of the cell</span></li></ul></td></tr><tr class="c17"><td class="c18"><p class="c2 c0"><span class="c11">Input data</span></p><ul class="c14 lst-kix_4ed4y1s6lj5w-0"><li class="c7 c0"><span>raw sequence for prediction</span></li><li class="c7 c0"><span>raw sequence for the best n hits</span></li><li class="c7 c0"><span>hit length</span></li><li class="c7 c0"><span>start/end index of the high-scoring segment pairs (hsp) for hits and query</span></li></ul><p class="c2 c0 c6"><span></span></p></td></tr><tr class="c17"><td class="c18"><p class="c2 c0"><span class="c11">Class information:</span></p><ul class="c14 lst-kix_e5f9ugvogyec-0 start"><li class="c1 c0"><span>short_heade: Duplication</span></li><li class="c1 c0"><span>header: Duplication</span></li><li class="c1 c0"><span>description: Check whether there is a duplicated subsequence in the predicted gene by counting the hsp residue coverage of the prediction, for each hit. Meaning of the output displayed: P-value of the Wilcoxon test which test the distribution of hit average coverage against 1. P-values higher than 5% pass the validation test.</span></li><li class="c1 c0"><span>cli_name: dup</span></li></ul><p class="c2 c0 c6"><span></span></p></td></tr><tr class="c17"><td class="c18"><p class="c2 c0"><span class="c11">Workflow</span></p><ul class="c14 lst-kix_sjinkw26u0wp-0 start"><li class="c7 c0"><span>get raw sequences for the first n hits (if they have not been already calculated)</span></li><li class="c7 c0"><span>for each hsp</span></li><li class="c7 c0"><span>subtract the raw sequence of the prediction</span></li><li class="c7 c0"><span>subtract the raw sequence of the hsps</span></li><li class="c7 c0"><span>local align them</span></li><li class="c7 c0"><span>initialize a coverage array for every hit</span></li><li class="c7 c0"><span>pass through the alignment and count how many </span></li><li class="c7 c0"><span>trick: average values are rounded by 2 decimals</span></li><li class="c7 c0"><span>wilcoxon test against 1 -&gt; otherwise we have duplication </span></li><li class="c7 c0"><span>in case of nucleotides -&gt; translate the hsp with rf = 0 or 1</span></li></ul></td></tr><tr class="c17"><td class="c18"><p class="c2 c0"><span class="c11">Code [1]</span></p><p class="c2 c0"><span class="c5"># get the first n hits</span></p><p class="c2 c0"><span class="c5">less_hits = @hits[0..[n-1,@hits.length].min]</span></p><p class="c2 c0"><span class="c5">useless_hits = []</span></p><p class="c2 c0"><span class="c5">begin</span></p><p class="c2 c0"><span class="c5">&nbsp; # get raw sequences for less_hits</span></p><p class="c2 c0"><span class="c5">&nbsp; less_hits.map do |hit|</span></p><p class="c2 c0"><span class="c5">&nbsp; #get gene by accession number</span></p><p class="c2 c0"><span class="c5">&nbsp; if hit.raw_sequence == nil</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; if hit.type == :protein</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;hit.get_sequence_by_accession_no(hit.accession_no, &quot;protein&quot;)</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; else</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;hit.get_sequence_by_accession_no(hit.accession_no, &quot;nucleotide&quot;)</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; end</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; if hit.raw_sequence == &quot;&quot;</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;useless_hits.push(hit)</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; end</span></p><p class="c2 c0"><span class="c5">&nbsp; end</span></p><p class="c2 c0"><span class="c5">end</span></p><p class="c2 c0"><span class="c5">rescue Exception =&gt; error</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp;raise NoInternetError</span></p><p class="c2 c0"><span class="c5">end</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5">useless_hits.each{|hit| less_hits.delete(hit)}</span></p><p class="c2 c0"><span class="c5">averages = []</span></p><p class="c2 c0"><span class="c5">less_hits.each do |hit|</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;coverage = Array.new(hit.length_protein,0)</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;hit.hsp_list.each do |hsp|</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;# align subsequences from the hit and prediction that match (if it&#39;s the case)</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;if hsp.hit_alignment != nil and hsp.query_alignment != nil</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; hit_alignment = hsp.hit_alignment</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; query_alignment = hsp.query_alignment</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;else</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; # indexing in blast starts from 1</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; hit_local = hit.raw_sequence[hsp.hit_from-1..hsp.hit_to-1]</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; query_local = prediction.raw_sequence[hsp.match_query_from-1..hsp.match_query_to-1]</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; # in case of nucleotide prediction sequence translate into protein</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; # use translate with reading frame 1 because</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # to/from coordinates of the hsp already correspond to the</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; # reading frame in which the prediction was read to match this hsp</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; if @type == :nucleotide</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; s = Bio::Sequence::NA.new(query_local)</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;query_local = s.translate</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; end</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; # local alignment for hit and query</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; seqs = [hit_local, query_local]</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; begin</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;options = [&#39;--maxiterate&#39;, &#39;1000&#39;, &#39;--localpair&#39;, &#39;--quiet&#39;]</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;mafft = Bio::MAFFT.new(@mafft_path, options)</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;report = mafft.query_align(seqs)</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;raw_align = report.alignment</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;align = []</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;raw_align.each { |s| align.push(s.to_s) }</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;hit_alignment = align[0]</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;query_alignment = align[1]</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;rescue Exception =&gt; error</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;raise NoMafftInstallationError</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; end</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;end</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;# check multiple coverage</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;# for each hsp of the curent hit</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;# iterate through the alignment and count the matching residues</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[*(0 .. hit_alignment.length-1)].each do |i|</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;residue_hit = hit_alignment[i]</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;residue_query = query_alignment[i]</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;if residue_hit != &#39; &#39; and residue_hit != &#39;+&#39; and residue_hit != &#39;-&#39;</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; if residue_hit == residue_query</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; # indexing in blast starts from 1</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;idx = i + (hsp.hit_from-1) - hit_alignment[0..i].scan(/-/).length</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; coverage[idx] += 1</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; end</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;end</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;end</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp;end</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp;overlap = coverage.reject{|x| x==0}</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp;if overlap != []</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;averages.push(overlap.inject(:+)/(overlap.length + 0.0)).map{|x| x.round(2)}</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; end</span></p><p class="c2 c0"><span class="c5">end</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5"># if all hsps match only one time</span></p><p class="c2 c0"><span class="c5">if averages.reject{|x| x==1} == []</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp;@validation_report = DuplciationValidationOutput.new(1)</span></p><p class="c2 c0"><span class="c5">end</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5">pval = wilcox_test(averages)</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5">@validation_report = DuplciationValidationOutput.new(pval)</span></p><p class="c2 c0"><span class="c5">[1] </span><span class="c5 c12 c26"><a class="c31" href="https://github.com/monicadragan/GeneValidator/blob/rubygem/lib/genevalidator/validation_duplication.rb#L65">https://github.com/monicadragan/GeneValidator/blob/rubygem/lib/genevalidator/validation_duplication.rb#L65</a></span><span class="c5">&nbsp;</span></p></td></tr><tr class="c17"><td class="c18"><p class="c2 c0"><span class="c11">Plots</span></p><p class="c2 c0"><span>No plots are generated</span></p></td></tr></tbody></table><p class="c0 c6"><span class="c11"></span></p><p class="c0 c6"><span></span></p><p class="c0"><span class="c11">4) Gene merge Validation</span></p><p class="c0 c6"><span></span></p><a href="#" name="7bfbd24e409e08175d370337aeeb4aa5452612b6"></a><a href="#" name="4"></a><table cellpadding="0" cellspacing="0" class="c30"><tbody><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Error cause</span></p><ul class="c14 lst-kix_gr4a4jnxityy-0 start"><li class="c1 c0"><span>gene prediction error (2 or more genes were merged together)</span></li></ul><p class="c2 c0 c6"><span></span></p></td></tr><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Input data</span></p><ul class="c14 lst-kix_4ed4y1s6lj5w-0"><li class="c7 c0"><span>start/end index of the high-scoring segment pairs (hsp) for hits and query</span></li></ul></td></tr><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Class information:</span></p><ul class="c14 lst-kix_cmqbbecr9wqr-0 start"><li class="c1 c0"><span>short_header:Gene_Merge(slope)</span></li><li class="c1 c0"><span>header: Gene Merge</span></li><li class="c1 c0"><span>description: Check whether BLAST hits make evidence about a merge of two genes that match the predicted gene. Meaning of the output displayed: slope of the linear regression of the relationship between the start and stop offsets of the hsps (see the plot). Invalid slopes are around 45 degrees.</span></li><li class="c1 c0"><span>cli_name: merge</span></li></ul><p class="c2 c0 c6"><span></span></p></td></tr><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Workflow</span></p><ul class="c14 lst-kix_myo05blafbk-0 start"><li class="c1 c0"><span>plot a 2D graph using the start/end offsets of the high-scoring segments in the prediction that match the hits (due to blast local alignment)</span></li><li class="c1 c0"><span>draw a line obtained by linear regression </span></li><li class="c1 c0"><span>the prediction is a merge of two genes if the slope of the line is between 0.4 and 1.2 (these thresholds are obtained empirically, after observing the slope of a large amount of queries). An image depicting this is available just below:</span></li></ul><p class="c2 c0 c6"><span></span></p><p class="c2 c0 c6"><span></span></p><p class="c0 c2"><img height="388" src="images/image03.png" width="609"></p><p class="c2 c0 c6"><span></span></p></td></tr><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Code [1]</span></p><p class="c2 c0 c6"><span class="c11"></span></p><p class="c2 c0"><span class="c9">lm_slope = slope[1]</span></p><p class="c2 c0"><span class="c9">y_intercept = slope[0]</span></p><p class="c2 c0 c6"><span class="c9"></span></p><p class="c2 c0"><span class="c9">@validation_report = GeneMergeValidationOutput.new(lm_slope)</span></p><p class="c2 c0 c6"><span class="c9"></span></p><p class="c2 c0"><span class="c9">plot1 = plot_2d_start_from(lm_slope, y_intercept)</span></p><p class="c2 c0"><span class="c9">@validation_report.plot_files.push(plot1)</span></p><p class="c2 c0"><span class="c9">plot2 = plot_matched_regions</span></p><p class="c2 c0"><span class="c9">@validation_report.plot_files.push(plot2)</span></p><p class="c2 c0"><span class="c9">@running_time = Time.now - start</span></p><p class="c2 c0"><span class="c9">return @validation_report</span></p><p class="c2 c0"><span class="c5">[1] </span><span class="c5 c12 c26"><a class="c31" href="https://github.com/monicadragan/GeneValidator/blob/rubygem/lib/genevalidator/validation_gene_merge.rb#L82">https://github.com/monicadragan/GeneValidator/blob/rubygem/lib/genevalidator/validation_gene_merge.rb#L82</a></span><span class="c5">&nbsp;</span></p></td></tr><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Plots</span></p><ul class="c14 lst-kix_whzqgipsg1bt-0 start"><li class="c1 c0"><span>scatterplot representing a 2D plot of the start/end offsets of the high-scoring segments in the prediction that match the hits</span></li></ul><p class="c2 c13 c0"><img height="282" src="images/image10.png" width="312"></p><p class="c2 c13 c0 c6"><span></span></p><ul class="c14 lst-kix_3kgcjgrfg4nh-0 start"><li class="c1 c0"><span>line plot representing the regions in the prediction that are matched by each hit</span></li></ul><p class="c2 c0 c6"><span></span></p><p class="c2 c13 c0"><img height="244" src="images/image04.png" width="312"></p><p class="c10 c2 c0 c6"><span></span></p></td></tr></tbody></table><p class="c0"><span class="c11">5) Validation based on multiple alignment - applicable only for protein sequences (for now)</span></p><p class="c0 c6"><span></span></p><a href="#" name="6b7c67f746414756386bf46cb6d5929d4a8fcd68"></a><a href="#" name="5"></a><table cellpadding="0" cellspacing="0" class="c30"><tbody><tr class="c17"><td class="c18"><p class="c2 c0"><span class="c11">Error causes</span></p><ul class="c14 lst-kix_hzkhz4hhlr8l-0 start"><li class="c1 c0"><span>sequencing error</span></li></ul></td></tr><tr class="c17"><td class="c18"><p class="c2 c0"><span class="c11">Input data</span></p><ul class="c14 lst-kix_4ed4y1s6lj5w-0"><li class="c7 c0"><span>raw sequence for prediction</span></li><li class="c7 c0"><span>raw sequence for the best n hits</span></li></ul><p class="c2 c0 c6"><span></span></p></td></tr><tr class="c17"><td class="c18"><p class="c2 c0"><span class="c11">Class information:</span></p><ul class="c14 lst-kix_4tq1geabcxa-0 start"><li class="c1 c0"><span>short_header = &quot;MA&quot;</span></li><li class="c1 c0"><span>header = &quot;Missing/Extra sequences&quot;</span></li><li class="c1 c0"><span>description = &quot;Finds missing and extra sequences in the prediction, based on the multiple alignment of the best hits. Meaning of the output displayed: the percentages of the missing/extra sequences with respect to the multiple alignment. Validation fails if one of these values is higher than 20%</span></li><li class="c1 c0"><span>cli_name = &quot;align&quot;</span></li></ul><p class="c2 c0 c6"><span></span></p></td></tr><tr class="c17"><td class="c18"><p class="c2 c0"><span class="c11">Workflow</span></p><ul class="c14 lst-kix_olh7ea290trr-0 start"><li class="c7 c0"><span>get raw sequences for the first n hits (if they have not been already calculated)</span></li><li class="c7 c0"><span>multiple align the hits with the prediction (mafft)</span></li><li class="c7 c0"><span>compute the pssm profile for the hits</span></li><li class="c7 c0"><span>remove isolated residues from the hits and from the prediction</span></li><li class="c7 c0"><span>compute gap coverage: count gaps that are in the prediction but not in the statistical model (see the Figure 1)</span></li><li class="c7 c0"><span>extra sequences: count residues that are in the prediction but not in the statistical model (see the Figure)</span></li><li class="c7 c0"><span>conserved regions: count the no of the residues conserved among the hits that appear in the prediction </span></li></ul><p class="c0 c6"><span></span></p><p class="c13 c0"><img height="105" src="images/image12.png" width="532"></p><p class="c13 c0"><span>(Figure 1)</span></p><p class="c0 c6"><span></span></p></td></tr><tr class="c17"><td class="c18"><p class="c2 c0"><span class="c11">Code [1]</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5"># get the first n hits</span></p><p class="c2 c0"><span class="c5">less_hits = @hits[0..[n-1,@hits.length].min]</span></p><p class="c2 c0"><span class="c5">useless_hits = []</span></p><p class="c2 c0"><span class="c5">begin</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; # get raw sequences for less_hits</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; less_hits.map do |hit|</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;if hit.raw_sequence == nil</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; #get gene by accession number</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; if hit.type == :protein</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;hit.get_sequence_by_accession_no(hit.accession_no, &quot;protein&quot;)</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; else</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;hit.get_sequence_by_accession_no(hit.accession_no, &quot;nucleotide&quot;)</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; end</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; if hit.raw_sequence == &quot;&quot;</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;useless_hits.push(hit)</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; end</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;end</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; end</span></p><p class="c2 c0"><span class="c5">end</span></p><p class="c2 c0"><span class="c5">useless_hits.each{|hit| less_hits.delete(hit)}</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5">begin</span></p><p class="c2 c0"><span class="c5">&nbsp; # multiple align sequences from &nbsp;less_hits with the prediction</span></p><p class="c2 c0"><span class="c5">&nbsp; multiple_align_mafft(prediction, less_hits)</span></p><p class="c2 c0"><span class="c5">&nbsp; rescue Exception =&gt; error</span></p><p class="c2 c0"><span class="c5">&nbsp; raise NoInternetError</span></p><p class="c2 c0"><span class="c5">end</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5">sm &nbsp;= get_sm_pssm(@multiple_alignment[0..@multiple_alignment.length-2])</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5"># remove isolated residues from the predicted sequence</span></p><p class="c2 c0"><span class="c5">prediction_raw = remove_isolated_residues(@multiple_alignment[@multiple_alignment.length-1])</span></p><p class="c2 c0"><span class="c5"># remove isolated residues from the statistical model</span></p><p class="c2 c0"><span class="c5">sm = remove_isolated_residues(sm)</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5">plot1 = plot_alignment(sm)</span></p><p class="c2 c0"><span class="c5">gaps = gap_validation(prediction_raw, sm)</span></p><p class="c2 c0"><span class="c5">extra_seq = extra_sequence_validation(prediction_raw, sm)</span></p><p class="c2 c0"><span class="c5">consensus = consensus_validation(prediction_raw, get_consensus(@multiple_alignment[0..@multiple_alignment.length-2]))</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5">@validation_report = AlignmentValidationOutput.new(gaps, extra_seq, consensus)</span></p><p class="c2 c0"><span class="c5">@validation_report.plot_files.push(plot1)</span></p><p class="c2 c0"><span class="c5">[1] </span><span class="c5 c12 c26"><a class="c31" href="https://github.com/monicadragan/GeneValidator/blob/rubygem/lib/genevalidator/validation_alignment.rb#L65">https://github.com/monicadragan/GeneValidator/blob/rubygem/lib/genevalidator/validation_alignment.rb#L65</a></span><span class="c5">&nbsp;</span></p><p class="c2 c0 c6"><span class="c5"></span></p></td></tr><tr class="c17"><td class="c18"><p class="c2 c0"><span class="c11">Plots</span></p><ul class="c14 lst-kix_yghjnj8s9xuc-0 start"><li class="c1 c0"><span>line plot representing the multiple alignment, the prediction and the statistical model (colors highlight gaps and extra sequences and consensus regions among the hits)</span></li></ul><p class="c2 c0 c6"><span></span></p><p class="c2 c13 c0"><img height="242" src="images/image05.png" width="307"></p></td></tr></tbody></table><p class="c0 c6"><span class="c11"></span></p><p class="c0"><span class="c11">6) Blast Reading Frame Validation - applicable only for nucleotide sequences</span></p><p class="c0 c6"><span></span></p><a href="#" name="45f3a72bbea95986471be06213168e0cbe3283d2"></a><a href="#" name="6"></a><table cellpadding="0" cellspacing="0" class="c30"><tbody><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Error causes</span></p><ul class="c14 lst-kix_vqtjz6k7exn7-0 start"><li class="c1 c0"><span>frame shift sequencing errors: various reading frames of the same sign (e.g single nucleotide was accidentally inserted, sequences are read in the same direction). </span></li><li class="c1 c10 c0"><span>various reading frames of different sign may still highlight a problem in the prediction, not a reading frame shift, but a possible merge between two genes (and this case is treated by the gene merge validation)</span><span class="c11">&nbsp;When reading frame is consistent for all the hits, we consider 3 cases: &nbsp;</span></li><li class="c1 c10 c0 c6"><span class="c11"></span></li></ul><ul class="c14 lst-kix_vqtjz6k7exn7-1 start"><li class="c22 c2 c0"><span class="c11">One direction gene merge (----&gt; -----------&gt;) &nbsp; - same sign reading frames for all hits</span></li><li class="c22 c2 c0"><span class="c11">Opposite direction gene merge (----&gt;&lt;----------) &nbsp; - two opposite sign reading frames for the hits</span></li><li class="c22 c2 c0"><span class="c11">otherwise we have no merge</span></li></ul><p class="c2 c0 c6"><span class="c11"></span></p></td></tr><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Input data</span></p><ul class="c14 lst-kix_4ed4y1s6lj5w-0"><li class="c7 c0"><span>reading frame provided by the blast output</span></li></ul></td></tr><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Class information:</span></p><ul class="c14 lst-kix_48dltnwvsqfp-0 start"><li class="c22 c2 c0"><span>short_header: Frame</span></li><li class="c22 c2 c0"><span>header: Reading Frame</span></li><li class="c22 c2 c0"><span>description: Check whether there is a single reading frame among BLAST hits. Otherwise there might be a reading frame shift in the query sequence. Meaning of the output displayed: (reading frame: no hsps)</span></li><li class="c22 c2 c0"><span>cli_name: frame</span></li></ul><p class="c2 c0 c6"><span></span></p></td></tr><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Workflow</span></p><ul class="c14 lst-kix_kmsnp6mmp404-0 start"><li class="c1 c0"><span>blastx translates each query in each of the six reading frames, and all the resulting six-protein sequences are compared to data in the databases (there&#39;s no evidence where the codon starts from, so all the 3 possibilities of starting the gene translation are considered in BLASTX; also another 3 possibilities when reading the sequence in reverse). </span></li></ul><ul class="c14 lst-kix_508olx789l51-0 start"><li class="c1 c0"><span>for all the hsps of each hit we check the reading frame number</span></li><li class="c1 c0"><span>store in a hash table the number of occurrences for each reading frame</span></li><li class="c1 c0"><span>when we count more reading frames of the same sign, there must have occurred a frame shift during the sequencing process. </span></li></ul></td></tr><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Code [1]</span></p><p class="c2 c0 c6"><span class="c11"></span></p><p class="c2 c0"><span class="c9">if type.to_s != &quot;nucleotide&quot;</span></p><p class="c2 c0"><span class="c9">&nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;@validation_report = ValidationReport.new(&quot;&quot;, :unapplicable)</span></p><p class="c2 c0"><span class="c9">&nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;return @validation_report</span></p><p class="c2 c0"><span class="c9">end</span></p><p class="c2 c0 c6"><span class="c9"></span></p><p class="c2 c0"><span class="c9">rfs = &nbsp;lst.map{ |x| x.hsp_list.map{ |y| y.query_reading_frame}}.flatten</span></p><p class="c2 c0"><span class="c9">&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;frames_histo = Hash[rfs.group_by { |x| x }.map { |k, vs| [k, vs.length] }]</span></p><p class="c2 c0 c6"><span class="c9"></span></p><p class="c2 c0"><span class="c9"># get the main reading frame</span></p><p class="c2 c0"><span class="c9">main_rf = frames_histo.map{|k,v| v}.max</span></p><p class="c2 c0"><span class="c9">@prediction.nucleotide_rf = frames_histo.select{|k,v| v==main_rf}.first.first</span></p><p class="c2 c0 c6"><span class="c9"></span></p><p class="c2 c0"><span class="c9">@validation_report = BlastRFValidationOutput.new(frames_histo)</span></p><p class="c2 c0"><span class="c5">[1] </span><span class="c5 c12 c26"><a class="c31" href="https://github.com/monicadragan/GeneValidator/blob/rubygem/lib/genevalidator/validation_blast_reading_frame.rb#L69">https://github.com/monicadragan/GeneValidator/blob/rubygem/lib/genevalidator/validation_blast_reading_frame.rb#L69</a></span><span class="c5">&nbsp;</span></p></td></tr><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Plots</span></p><p class="c2 c0"><span>No plots are generated</span></p></td></tr></tbody></table><p class="c0 c6"><span class="c11"></span></p><p class="c0"><span class="c11">7) Open reading frame - applicable only for nucleotide sequences</span></p><p class="c0 c6"><span></span></p><a href="#" name="3dc6cbb6d28e74d814111cdc14a8761743d604f4"></a><a href="#" name="7"></a><table cellpadding="0" cellspacing="0" class="c30"><tbody><tr class="c17"><td class="c18"><p class="c2 c0"><span class="c11">Error causes</span></p><ul class="c14 lst-kix_g96zcqu3kh41-0 start"><li class="c1 c0"><span class="c5">many small ORF</span></li><li class="c1 c0"><span class="c5">frame shift</span></li></ul></td></tr><tr class="c17"><td class="c18"><p class="c2 c0"><span class="c11">Input data</span></p><ul class="c14 lst-kix_4ed4y1s6lj5w-0"><li class="c7 c0"><span>only query raw sequence is needed for this validation (got from the fasta file)</span></li></ul><p class="c2 c0 c6"><span></span></p></td></tr><tr class="c17"><td class="c18"><p class="c2 c0"><span class="c11">Class information:</span></p><ul class="c14 lst-kix_x7t5b7jbczs4-0 start"><li class="c1 c0"><span>short_header: ORF</span></li><li class="c1 c0"><span>header: Main ORF</span></li><li class="c1 c0"><span>description: Check whether there is a single main Open Reading Frame in the predicted gene. Applicable only for nucleotide queries. Meaning of the output displayed: %=MAIN ORF COVERAGE. Coverage higher than 80% passe the validation test.</span></li><li class="c1 c0"><span>cli_name: orf</span></li></ul><p class="c2 c0 c6"><span></span></p></td></tr><tr class="c17"><td class="c18"><p class="c2 c0"><span class="c11">Workflow</span></p><ul class="c14 lst-kix_g3lepnw8n0x6-0 start"><li class="c1 c0"><span>for each reading frame get all the ORFs, with customizable stop and start codons</span></li><li class="c1 c0"><span>check if there is any ORF read in a certain reading frame that il longer than 80% of the length of the prediction </span></li></ul><p class="c2 c0 c6"><span></span></p></td></tr><tr class="c17"><td class="c18"><p class="c2 c0"><span class="c11">Code</span><span class="c11">&nbsp;[1]</span></p><p class="c2 c0"><span class="c5">if type.to_s != &quot;nucleotide&quot;</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;@validation_report = ValidationReport.new(&quot;&quot;, :unapplicable)</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;return @validation_report</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;end</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;orfs = get_orfs</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;# check if longest ORF / prediction &gt; 0.8 (ok)</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;prediction_len = prediction.raw_sequence.length</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;longest_orf = orfs.map{|elem| elem[1].map{|orf| orf[1]-orf[0]}}.flatten.max</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ratio = &nbsp;longest_orf/(prediction_len + 0.0)</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;len = @prediction.raw_sequence.length</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;f = File.open(&quot;#{@filename}_orfs.json&quot; , &quot;w&quot;)</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;lst = @hits.sort{|a,b| a.length&lt;=&gt;b.length}</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;f.write((orfs.each_with_index.map{|elem, i| {&quot;y&quot;=&gt;elem[0], &quot;start&quot;=&gt;0, &quot;stop&quot;=&gt;len, &quot;color&quot;=&gt;&quot;black&quot;}} +</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;orfs.each_with_index.map{|elem, i| elem[1].map{|orf| {&quot;y&quot;=&gt;elem[0], &quot;start&quot;=&gt;orf[0], &quot;stop&quot;=&gt;orf[1], &quot;color&quot;=&gt;&quot;red&quot;}}}.flatten).to_json)</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;f.close</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;@plot_files.push(Plot.new(&quot;#{@filename}_orfs.json&quot;.scan(/\/([^\/]+)$/)[0][0],</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;:lines,</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&quot;Open reading frame with START codon&quot;,</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&quot;&quot;,</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&quot;length&quot;,</span></p><p class="c2 c0"><span class="c5">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&quot;Reading Frame&quot;))</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5">@validation_report = ORFValidationOutput.new(orfs, ratio)</span></p><p class="c2 c0"><span class="c5">[1] </span><span class="c5 c12 c26"><a class="c31" href="https://github.com/monicadragan/GeneValidator/blob/rubygem/lib/genevalidator/validation_open_reading_frame.rb#L76">https://github.com/monicadragan/GeneValidator/blob/rubygem/lib/genevalidator/validation_open_reading_frame.rb#L76</a></span><span class="c5">&nbsp;</span></p></td></tr><tr class="c17"><td class="c18"><p class="c2 c0"><span class="c11">Plots</span></p><p class="c2 c0 c6"><span></span></p><ul class="c14 lst-kix_l7t8lma3v5dd-0 start"><li class="c1 c0"><span>line plot depicting the ORFs for each reading frame</span></li></ul><p class="c2 c0 c6"><span></span></p><p class="c2 c13 c0"><img height="377" src="images/image06.png" width="361"></p></td></tr></tbody></table><p class="c0 c6"><span></span></p><p class="c0"><span class="c11">8</span><span class="c11">) Codon usage - this is future work</span></p><p class="c0 c6"><span></span></p><a href="#" name="0f629552535ca45d1ec71e0d3086a0eb88fa79b1"></a><a href="#" name="8"></a><table cellpadding="0" cellspacing="0" class="c30"><tbody><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Error causes</span></p><ul class="c14 lst-kix_ao4z72tcnso6-0 start"><li class="c1 c0"><span>the gene incorrectly includes an intron</span></li></ul><ul class="c14 lst-kix_vz631vhbp5pc-0 start"><li class="c1 c0"><span>in the sequencing process the gene was infested with other substance or even with bacterial or human DNA</span></li></ul></td></tr><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Input data</span></p><ul class="c14 lst-kix_4ed4y1s6lj5w-0"><li class="c7 c0"><span>query raw sequence (got from the fasta file)</span></li></ul></td></tr><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Class information:</span></p><p class="c2 c0 c6"><span></span></p></td></tr><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Workflow</span></p><p class="c2 c0 c6"><span></span></p></td></tr><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Code</span></p><p class="c2 c0 c6"><span class="c5"></span></p></td></tr><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Plots</span></p><p class="c2 c0"><span>No plots are generated</span></p></td></tr></tbody></table><p class="c0"><span class="c11 c21">Running times:</span></p><p class="c0 c6"><span></span></p><a href="#" name="340bbf72f52067dfcada94f02429dfe3440edbf6"></a><a href="#" name="9"></a><table cellpadding="0" cellspacing="0" class="c30"><tbody><tr class="c17"><td class="c23"><p class="c2 c0"><span>Validation</span></p></td><td class="c24"><p class="c2 c0"><span>10 q</span></p></td><td class="c28"><p class="c2 c0"><span>30 q</span></p></td><td class="c19"><p class="c2 c0"><span>1,000 q</span></p></td><td class="c20"><p class="c2 c0"><span>10,000 q</span></p></td><td class="c16"><p class="c2 c0"><span>Time consumers</span></p></td></tr><tr><td class="c23"><p class="c2 c0"><span>Length Cluster Validation</span></p></td><td class="c24"><p class="c2 c0"><span>0.104s</span></p></td><td class="c28"><p class="c2 c0"><span>0.025s</span></p></td><td class="c19"><p class="c2 c0"><span>-</span></p></td><td class="c20"><p class="c2 c0"><span>-</span></p></td><td class="c16"><p class="c2 c0 c6"><span></span></p></td></tr><tr class="c17"><td class="c23"><p class="c2 c0"><span>Length Rank Validation</span></p></td><td class="c24"><p class="c2 c0"><span>0.00s</span></p></td><td class="c28"><p class="c2 c0"><span>0.00s</span></p></td><td class="c19"><p class="c2 c0"><span>-</span></p></td><td class="c20"><p class="c2 c0"><span>-</span></p></td><td class="c16"><p class="c2 c0 c6"><span></span></p></td></tr><tr class="c17"><td class="c23"><p class="c2 c0"><span>Gene_Merge</span></p></td><td class="c24"><p class="c2 c0"><span>0.003s</span></p></td><td class="c28"><p class="c2 c0"><span>0.01s</span></p></td><td class="c19"><p class="c2 c0"><span>-</span></p></td><td class="c20"><p class="c2 c0"><span>-</span></p></td><td class="c16"><p class="c2 c0 c6"><span></span></p></td></tr><tr><td class="c23"><p class="c2 c0"><span>Duplication Validation</span></p></td><td class="c24"><p class="c2 c0"><span>9.458s</span></p></td><td class="c28"><p class="c2 c0"><span>10.488s</span></p></td><td class="c19"><p class="c2 c0"><span>-</span></p></td><td class="c20"><p class="c2 c0"><span>-</span></p></td><td class="c16"><ul class="c14 lst-kix_fscfg01f3awm-0 start"><li class="c1 c0"><span>potential use of the webservice to get raw sequences</span></li><li class="c1 c0"><span>MAFFT local alignment</span></li></ul></td></tr><tr class="c17"><td class="c23"><p class="c2 c0"><span>Multiple Alignment Validation</span></p></td><td class="c24"><p class="c2 c0"><span>1.785s</span></p><p class="c2 c0 c6"><span></span></p></td><td class="c28"><p class="c2 c0"><span>0.645s</span></p></td><td class="c19"><p class="c2 c0"><span>-</span></p></td><td class="c20"><p class="c2 c0"><span>-</span></p></td><td class="c16"><ul class="c14 lst-kix_dpynbghv8b9j-0 start"><li class="c1 c0"><span>potential use of the webservice to get raw sequences (if was not retrieved in a previous validation)</span></li><li class="c1 c0"><span>MAFFT multiple alignment</span></li></ul></td></tr><tr class="c17"><td class="c23"><p class="c2 c0"><span>Frame Validation</span></p></td><td class="c24"><p class="c2 c0"><span>0.0s</span></p></td><td class="c28"><p class="c2 c0"><span>0.0s</span></p></td><td class="c19"><p class="c2 c0"><span>-</span></p></td><td class="c20"><p class="c2 c0"><span>-</span></p></td><td class="c16"><p class="c2 c0 c6 c34"><span></span></p></td></tr><tr class="c17"><td class="c23"><p class="c2 c0"><span>ORF Validation</span></p></td><td class="c24"><p class="c2 c0"><span>0.015s</span></p></td><td class="c28"><p class="c2 c0"><span>0.012s</span></p></td><td class="c19"><p class="c2 c0"><span>-</span></p></td><td class="c20"><p class="c2 c0"><span>-</span></p></td><td class="c16"><p class="c34 c2 c0 c6"><span></span></p></td></tr></tbody></table><p class="c0 c6"><span></span></p><p class="c13 c0 c6"><span class="c8"></span></p><hr style="page-break-before:always;display:none;"><p class="c13 c0 c6"><span class="c8"></span></p><p class="c13 c0"><span class="c8">C. What&rsquo;s the structure of the code</span></p><p class="c10 c0 c6"><span class="c11"></span></p><p class="c13 c0 c6"><span class="c11 c21"></span></p><p class="c13 c0 c6"><span class="c11 c21"></span></p><p class="c13 c0"><img height="612" src="images/image11.jpg" width="919"></p><p class="c13 c0 c6"><span class="c11 c21"></span></p><hr style="page-break-before:always;display:none;"><p class="c13 c0 c6"><span class="c11 c21"></span></p><p class="c13 c0"><span class="c8">D. How to run the tool</span></p><p class="c0"><span class="c11 c15">1. Prerequisite:</span></p><p class="c0 c6"><span></span></p><ul class="c14 lst-kix_cbmcswhr1ftl-0 start"><li class="c7 c0"><span class="c11">Ruby</span><span>&nbsp;(&gt;= 1.9.3)</span></li><li class="c7 c0"><span class="c11">R</span><span>&nbsp;(&gt;= 2.14.2)</span></li><li class="c7 c0"><span class="c11">RubyGems</span><span>&nbsp;(&gt;= 1.3.6)</span></li><li class="c7 c0"><span class="c11">NCBI BLAST+</span><span>&nbsp;(&gt;= 2.2.25+)</span></li><li class="c7 c0"><span class="c11">MAFFT installation</span><span>&nbsp;(download it from :</span><span><a class="c31" href="http://mafft.cbrc.jp/alignment/software/">&nbsp;</a></span><span class="c12 c26"><a class="c31" href="http://mafft.cbrc.jp/alignment/software/">http://mafft.cbrc.jp/alignment/software/</a></span><span>&nbsp;).</span></li></ul><p class="c0 c6"><span></span></p><p class="c0"><span class="c11 c15">2. Installation:</span></p><p class="c0 c6"><span class="c11 c15"></span></p><ol class="c14 lst-kix_9drbe3jtkkuh-0 start" start="1"><li class="c7 c0"><span>Get the source code</span></li></ol><p class="c0 c27"><span class="c9">$ git clone git@github.com:monicadragan/gene_prediction.git</span></p><ol class="c14 lst-kix_9drbe3jtkkuh-0"><li class="c7 c0"><span>Be sudo and build the gem</span></li></ol><p class="c27 c0"><span class="c9">$ sudo rake</span></p><ol class="c14 lst-kix_9drbe3jtkkuh-0"><li class="c7 c0"><span>Run GeneValidation</span></li></ol><p class="c27 c0"><span class="c9">$ genevalidator type [validations] [skip_blast] [start] [tabular] [mafft] FILE</span></p><p class="c27 c0 c6"><span></span></p><p class="c0"><span>Example that runs all validations on a set of ant gene predictions:</span></p><p class="c0"><span class="c9">$ genevalidator -x data/prot_Solenopsis.xml data/prot_Solenopsis.fa</span></p><p class="c0 c6"><span></span></p><p class="c0"><span>To learn more:</span></p><p class="c0"><span class="c9">$ genevalidator -h</span></p><p class="c0 c6"><span class="c9"></span></p><p class="c0"><span class="c11 c15">3. Input</span></p><p class="c0 c6"><span class="c11 c15"></span></p><ul class="c14 lst-kix_xc244bhq3ube-0 start"><li class="c7 c0"><span class="c11">FASTA file</span><span>&nbsp;with the prediction queries (queries must be of the same type: </span><span class="c11">mRNA or proteins</span><span>. Mixed type FASTA files are not allowed!).</span></li><li class="c7 c0"><span>[optional] </span><span class="c11">Precalculated BLAST output file</span><span>&nbsp;in xml format (-outfmt 5) or tabular format (-outfmt 6 or -outfmt 7 or &ldquo;Hit Table (CSV)&rdquo; in the online version of BLAST). Please specify the --tabular argument if you use tabular format input with nonstandard columns. Also, </span><span class="c11">we search only in protein database using a blastp for protein queries and blastx for nucleotide queries. </span></li></ul><p class="c0 c6"><span class="c11 c15"></span></p><p class="c0"><span class="c11 c15">4. Output</span></p><p class="c0 c6"><span class="c11 c15"></span></p><p class="c0"><span>By running GeneValidator on your dataset you get numbers and plots. Some relevant files will be generated at the same path with the input file. The results are available in 3 formats:</span></p><ul class="c14 lst-kix_jamjx369pdr8-0 start"><li class="c7 c0"><span class="c11">console table output</span></li><li class="c7 c0"><span>validation results in </span><span class="c11">YAML format</span><span>&nbsp;(the YAML file has the same name with the input file + YAML extension)</span></li><li class="c7 c0"><span class="c11">html output</span><span>&nbsp;with plot visualization (the useful files will be generated in the &#39;html&#39; directory, at the same path with the input file). Check the </span><span class="c11">index.html</span><span>&nbsp;file to see the output.</span></li></ul><p class="c13 c0 c6"><span></span></p><p class="c13 c0"><span class="c11 c25">! Note: for the moment check the html output with Firefox browser only !</span></p><p class="c13 c0"><span class="c11 c25">(because our d3js cannot be visualized from other browsers without a local webserver) </span></p><p class="c0 c6"><span class="c11 c15"></span></p><p class="c0 c6"><span></span></p><p class="c0"><span class="c11 c15">5. Scenarios</span></p><p class="c0 c6"><span class="c11 c15"></span></p><p class="c0"><span class="c11 c25">5.1</span><span class="c25">&nbsp;Quickly test the tool or validate a small set of predictions </span><span class="c11 c25">(internet connection required).</span></p><p class="c13 c0"><span class="c9">$ genevalidator data/prot_Solenopsis.fa</span></p><p class="c0 c6"><span class="c11"></span></p><p class="c0"><span class="c29">What happens backstage: blast hits and raw sequences are retrieved from remote databases and analysed for each </span><span class="c29">query</span><span class="c29">.</span></p><p class="c0 c6"><span class="c11"></span></p><p class="c10 c0"><span class="c11 c25">5.2</span><span>&nbsp;</span><span class="c25">Big input files (&gt;100 queries) &amp; you have access to a more powerful machine</span><span>&nbsp;that pre-calculates overnight the blast output file. </span><span class="c11">Use blastp for proteins and blastx for nucleotide queries.</span></p><p class="c10 c0 c6"><span class="c11"></span></p><p class="c0"><span class="c11">5.2.1 If </span><span>you have time and a really powerful machine you can </span><span class="c11">take the blast xml output</span><span>. This way you will save time when running GeneValidator (because you precalculate local alignments with BLAST) &nbsp;</span></p><p class="c0 c6"><span></span></p><p class="c0"><span>Example: </span></p><p class="c0 c6"><span></span></p><p class="c0"><span># take blast output for </span><span class="c11">protein queries </span><span>(for </span><span class="c11">mrna queries </span><span>use </span><span class="c11">blastx</span><span>)</span></p><p class="c0"><span class="c9">$ blastp -query predictions.fasta -db /path/to/db/nr/nr -outfmt 5 -max_target_seqs 200 -num_threads 48 -out predictions_blast.xml</span></p><p class="c0 c6"><span></span></p><p class="c0"><span># run validations</span></p><p class="c0"><span class="c9">$ genevalidator -x predictions_blast.xml predictions.fasta</span></p><p class="c0 c6"><span class="c9"></span></p><p class="c0"><span class="c11">5</span><span class="c11">.2.2 Take blast tabular output</span><span>&nbsp;(customized so that you have all the necessary columns -- see the table with data needed for validations in section B)</span></p><p class="c0 c6"><span></span></p><p class="c0"><span>Example: </span></p><p class="c0"><span># take blast output for protein queries</span></p><p class="c0"><span class="c9">$ blastp -query predictions.fasta -db /path/to/db/nr/nr -outfmt &quot;6 qseqid sseqid sacc slen qstart qend sstart send length qframe pident evalue&quot; -max_target_seqs 200 -num_threads 48 -out predictions_blast.tab</span></p><p class="c0 c6"><span></span></p><p class="c0"><span># run validations</span></p><p class="c0"><span class="c9">$ genevalidator -x predictions_blast.tab -t &quot;qseqid sseqid sacc slen qstart qend sstart send length qframe pident evalue&quot; predictions.fasta </span></p><p class="c0 c6"><span class="c9"></span></p><p class="c0"><span class="c11 c9">5.3</span><span class="c9">&nbsp;You are interested in </span><span class="c11 c9">certain validations</span><span class="c9">&nbsp;only (let&rsquo;s take length validations and duplication check). Look up the CLI names for these validations in the table (Section B) and add them as argument so that GeneValidator will process those 3 only.</span></p><p class="c0 c6"><span class="c9"></span></p><p class="c0"><span class="c9">$ genevalidator -x predictions_blast.xml predictions.fasta -v &lsquo;lenc lenr dup&rsquo;</span></p><p class="c0 c6"><span class="c9"></span></p><p class="c0"><span class="c11 c9">5.4 Start validating from a certain query in the fasta file: </span><span class="c9">the following command skips the first 9 queries</span></p><p class="c0 c6"><span class="c9"></span></p><p class="c0"><span class="c9">$ genevalidator -x predictions_blast.xml predictions.fasta -s 10</span></p><p class="c0 c6"><span class="c9"></span></p><p class="c0 c6"><span class="c9"></span></p><hr style="page-break-before:always;display:none;"><p class="c0 c6"><span class="c9"></span></p><p class="c13 c0"><span class="c8">E. Write a new validation</span></p><a href="#" name="9637c6f5e7a1968f06c01e5e4f2d57ec1587dbef"></a><a href="#" name="10"></a><table cellpadding="0" cellspacing="0" class="c30"><tbody><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Add validation information</span></p><ul class="c14 lst-kix_4ed4y1s6lj5w-0"><li class="c7 c0"><span>short description - may not contain spaces</span></li><li class="c7 c0"><span>expected result</span></li></ul><p class="c2 c0 c6"><span></span></p></td></tr><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Parent classes (must be extended)</span></p><p class="c10 c2 c0 c6"><span></span></p><p class="c10 c2 c0"><span>Each validation is defined by 2 extended classes:<br></span><span class="c11">(1) extend ValidatioReport [1]</span><span>&nbsp;- stores the output of the validation test and some methods used for data representation, that may need to be overridden:</span></p><ul class="c14 lst-kix_owpom8h3nkwo-0 start"><li class="c1 c0"><span class="c11">validation</span><span>: returns oe of the 4 results: </span></li></ul><ul class="c14 lst-kix_owpom8h3nkwo-1 start"><li class="c22 c2 c0"><span>:yes </span></li><li class="c22 c2 c0"><span>:no </span></li><li class="c2 c0 c22"><span>:unapplicable </span></li><li class="c22 c2 c0"><span>:error</span></li></ul><ul class="c14 lst-kix_owpom8h3nkwo-0"><li class="c1 c10 c0"><span class="c11">print</span><span>: returns the output that will be displayed in the outputs</span></li><li class="c1 c10 c0"><span class="c11">color</span><span>: based on the validation result returns the background color for the corresponding cell in the html output (at the moment the possible values are &ldquo;success&rdquo;, &ldquo;warning&rdquo; and &ldquo;danger&rdquo; -- which are classes in the CSS file)</span></li></ul><p class="c2 c0 c6"><span></span></p><p class="c10 c2 c0"><span>Initialize internal variables (which initially have some default values):</span></p><ul class="c14 lst-kix_pdbqztwjr65g-0 start"><li class="c1 c10 c0"><span class="c11">expected</span><span>&nbsp;: the expected correct result for the validation. The test will pass if the expected result is equal to the validation result (:yes or :no)</span></li><li class="c1 c10 c0"><span>initialize </span><span class="c11">plot_files</span><span>&nbsp;with the empty Array in case you will generate plots (by default this variable is nil)</span></li><li class="c1 c10 c0"><span>initialize </span><span class="c11">errors</span><span>&nbsp;with the empty Array in case you will handle and display in the output the possible errors of this validation</span></li></ul><p class="c10 c2 c0"><span><br></span><span class="c11">(2) extend ValidationTest [2] </span><span>(the &#39;run&#39; method must be overloaded, run must return ValidationReport class type, &nbsp;some fields have to be updated: the validation name used in the header, description, plot_files)</span></p><p class="c2 c0 c6"><span class="c11"></span></p><p class="c10 c2 c0"><span>Initialize internal variables (which initially have some default values):</span></p><ul class="c14 lst-kix_48dltnwvsqfp-0"><li class="c32 c2 c0"><span class="c11">short_header</span><span>: will appear in the console output (may not contain spaces)</span></li><li class="c32 c2 c0"><span class="c11">header</span><span>: will appear in the header of the table - in html output </span></li><li class="c2 c0 c32"><span class="c11">description</span><span>: short description of the validation -- will appear in the html output</span></li><li class="c32 c2 c0"><span class="c11">cli_name</span><span>: validation can be referred by this alias when selecting the validations to be run (-v argument in the command line)</span></li><li class="c32 c2 c0"><span class="c11">running_time</span><span>: if you want to make overall running time statistics</span></li></ul><p class="c2 c0"><span>[1] </span><span class="c12 c26"><a class="c31" href="https://github.com/monicadragan/GeneValidator/blob/rubygem/lib/genevalidator/validation_output.rb">https://github.com/monicadragan/GeneValidator/blob/rubygem/lib/genevalidator/validation_output.rb</a></span><span>&nbsp;</span></p><p class="c2 c0"><span>[2] </span><span class="c12 c26"><a class="c31" href="https://github.com/monicadragan/GeneValidator/blob/rubygem/lib/genevalidator/validation_test.rb">https://github.com/monicadragan/GeneValidator/blob/rubygem/lib/genevalidator/validation_test.rb</a></span><span>&nbsp;</span></p></td></tr><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Add the new validation to the tool</span></p><p class="c0 c35"><span>Add the validation to the validations list, which is further processed for yaml/html/console visualization<br><br></span><span class="c11">Code [1]:</span></p><p class="c10 c0"><span class="c9"># context code #</span></p><p class="c10 c0"><span class="c9">validations = []</span></p><p class="c10 c0"><span class="c9">validations.push LengthClusterValidation.new(@type, prediction, hits, plot_path, plots)</span></p><p class="c10 c0"><span class="c9">&nbsp;&hellip;</span></p><p class="c10 c0"><span class="c11 c33 c9">## what you need to add is this line that adds your validation class to the list of validations. Your class must extend ValidationTest ##</span></p><p class="c10 c0 c6"><span class="c11 c33 c9"></span></p><p class="c10 c0"><span class="c11 c9 c33">validations.push YourValidationClass(type, prediction, hits, other_arguments)</span></p><p class="c10 c0 c6"><span class="c9"></span></p><p class="c10 c0"><span class="c9"># context code #</span></p><p class="c10 c0"><span class="c9"># check the class type of the elements in the list</span></p><p class="c10 c0"><span class="c9"># this will raise an error if YourValidationClass does not extend ValidationTest validations.map do |v|</span></p><p class="c10 c0"><span class="c9">&nbsp; raise ValidationClassError unless v.is_a? ValidationTest</span></p><p class="c10 c0"><span class="c9">end</span></p><p class="c10 c0 c6"><span class="c9"></span></p><p class="c10 c0"><span class="c9"># run validations</span></p><p class="c10 c0"><span class="c9">validations.map{|v| v.run}</span></p><p class="c10 c0 c6"><span class="c9"></span></p><p class="c10 c0"><span class="c9"># check the class type of the validation reports</span></p><p class="c10 c0"><span class="c9"># this will raise an error if the run method of YourValidationClass does not return ValidationReport</span></p><p class="c10 c0"><span class="c9">validations.map do |v|</span></p><p class="c10 c0"><span class="c9">&nbsp; raise ValidationClassError unless v.validation_report.is_a? ValidationReport</span></p><p class="c0 c10"><span class="c9">end</span></p><p class="c2 c0 c6"><span class="c5"></span></p><p class="c2 c0"><span class="c5">[1] </span><span class="c5 c12 c26"><a class="c31" href="https://github.com/monicadragan/GeneValidator/blob/rubygem/lib/genevalidator/validation.rb#L284">https://github.com/monicadragan/GeneValidator/blob/rubygem/lib/genevalidator/validation.rb#L284</a></span><span class="c5">&nbsp;</span></p></td></tr><tr class="c17"><td class="c4"><p class="c2 c0"><span class="c11">Add plots</span></p><p class="c2 c0 c6"><span class="c11"></span></p><p class="c2 c0"><span class="c11">A. Generate json file according to the type of plot</span></p><p class="c2 c0 c6"><span class="c11"></span></p><p class="c2 c0"><span class="c11">1) (clustered) bars plot</span><span>, main cluster is in red: [ [{&quot;key&quot;:109,&quot;value&quot;:1,&quot;main&quot;:false}], [{&quot;key&quot;:122,&quot;value&quot;:32,&quot;main&quot;:true},{&quot;key&quot;:123,&quot;value&quot;:33,&quot;main&quot;:true}], [...]]</span></p><p class="c2 c0 c6"><span class="c11"></span></p><p class="c2 c0"><span class="c11">2) scatter plot</span><span>: </span></p><p class="c2 c0"><span>[{&quot;x&quot;:1,&quot;y&quot;:626},{&quot;x&quot;:1,&quot;y&quot;:626}...]</span></p><p class="c2 c0 c6"><span></span></p><p class="c2 c0"><span class="c11">3) lines</span><span>: [{&quot;y&quot;:1,&quot;start&quot;:0,&quot;stop&quot;:356,&quot;color&quot;:&quot;black&quot;},{&quot;y&quot;:1,&quot;start&quot;:40,&quot;stop&quot;:200,&quot;color&quot;:&quot;red&quot;}...]</span></p><p class="c2 c0 c6"><span></span></p><p class="c2 c0"><span class="c11">B. Create a plot object, according to the plot type </span></p><p class="c2 c0"><span class="c9">&nbsp; &nbsp;plot1 = Plot.new(filemane,</span></p><p class="c2 c0"><span class="c9">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;plot_type # which can be </span><span class="c11">:scatter, :line or :bar</span></p><p class="c2 c0"><span class="c9">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;plot_title,</span></p><p class="c2 c0"><span class="c9">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;legend**,</span></p><p class="c2 c0"><span class="c9">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;xTitle,</span></p><p class="c2 c0"><span class="c9">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;yTitle,</span></p><p class="c2 c0"><span class="c9">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;aux_parameter1*,</span></p><p class="c2 c0"><span class="c9">&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;aux_parameter2*)</span></p><p class="c2 c0 c6"><span class="c9"></span></p><p class="c2 c0"><span class="c11 c9">** legend is a String that _must_ have the following format:</span></p><p class="c2 c0"><span class="c11 c9">&quot;item, color1;item with spaces, color2&quot;</span></p><p class="c2 c0"><span class="c11 c9">* auxiliary parameters</span><span class="c9">&nbsp;were used by now for passing the length of prediction (in length histogram) or the slope of the regression line (gene merge validation) or the number of lines (line plots)</span></p><p class="c2 c0 c6"><span class="c9"></span></p><p class="c2 c0"><span class="c11 c9">C. Add the plot object to the plot_files array</span></p><p class="c2 c0 c6"><span></span></p><p class="c2 c0"><span class="c9">@validation_report.plot_files.push(plot1)</span></p></td></tr></tbody></table><p class="c0 c6"><span class="c11 c21"></span></p><p class="c13 c0 c6"><span class="c11 c21"></span></p><hr style="page-break-before:always;display:none;"><p class="c13 c0 c6"><span class="c11 c21"></span></p><p class="c13 c0"><span class="c8">F. Output results</span></p><p class="c13 c0 c6"><span class="c8"></span></p><p class="c0"><span class="c8">Some validation outputs for protein and mRNA are available here:</span></p><p class="c0"><span class="c12 c26"><a class="c31" href="http://swarm.cs.pub.ro/~mdragan/gsoc2013/genevalidator">http://swarm.cs.pub.ro/~mdragan/gsoc2013/genevalidator</a></span><span>&nbsp; </span></p><p class="c0 c6"><span></span></p><p class="c0"><img height="360" src="images/image01.png" width="799"></p><p class="c0 c6"><span></span></p></body></html>